---
output:
    html_document:
        css: style.css
chunk_output_type: console
header-includes:
- \pagenumbering{gobble}
params:
    set_title: "Extended Data Figures"
title: "`r params$set_title`"
---


```{r setup, include = FALSE, echo = FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE, autodep = knitr::dep_prev())
source(here::here('r/prerequisites.R'))
```

# EDF 1

### EDF-1a

```{r panel_edf1a, width=6, height=7}

## cohort sample-type breakdown
d <- fread(here('data/data_clinical.txt.gz'))
tbl <- table.freq(d$metamaintype)
type_order <- c(tbl$value[tbl$value!='Other'],'Other')
tbl$metamaintype <- factor(tbl$value, levels=(type_order))
p <- ggplot(tbl,aes(x=metamaintype,y=N)) +
    geom_bar(stat='identity',color='white',fill='black',size=0.2) + 
    scale_y_continuous(expand=c(0,0)) +
    theme_std(base_size=14) +
    theme(axis.text.x=element_text(angle=90,hjust=1,vjust=0.5)) +
    labs(x=NULL,y='Number of samples') 
p

```

***

\pagebreak

### EDF-1b

```{r panel_edf1b}

## load data from permutations
perm <- readRDS(here::here('data/observed_vs_expected_composites_impact_precalculated.rds'))
obs <- perm$obs_all
exp <- median(perm$dat_all$prop)

## plot results
perm$dat$group <- 'All samples'
plot_dat <- perm$dat
plot_dat$data <- 'Randomly shuffled'
plot_dat$prop <- plot_dat$prop * 100
obs_dat <- data.table(group='All samples')
obs_dat$prop <- perm$obs*100
obs_dat$data <- 'Observed'
pd <- rbind(plot_dat, obs_dat, fill=T)
x <- break_axis(pd$prop,maxlower=16,minupper=22,lowerticksize=1,upperticksize=1,ratio_lower_to_upper=0.5)
pd$prop2 <- x$newy
expected_color <- c('black','red')
names(expected_color) <- c('Randomly shuffled','Observed')
prop_label <- paste0(prettyNum(obs_dat$prop,digits=3),'%')
p <- ggplot(pd,aes(x=prop2)) +
    geom_histogram(binwidth=0.1,aes(fill=data),color='white',size=0.2) +
    geom_vline(data=pd[data=='Observed',],aes(xintercept=prop2),color='red',linetype='dashed',size=0.5) +
    geom_text(data=pd[1,],x=17.75,y=3000,color='red',label=prop_label) +
    scale_fill_manual(values=expected_color,name=NULL) +
    scale_x_continuous(breaks=c(x$breaks,18), labels=c(x$labels,24), limits=c(13.8,18.1)) +
    scale_y_continuous(expand=c(0,0)) +
    labs(x='Percent of cases\n(MSK-IMPACT)',y='Frequency') + 
    theme_std(base_size=14) +
    theme(legend.position='none',axis.ticks.y=element_blank(),axis.text.y=element_blank())
p

```

***

\pagebreak

### EDF-1c

```{r panel_edf1c}

## note that we are correcting the number of samples overall to be the number of samples including those with 0 mutations based on the clinical data
info_all <- readRDS(here('data/observed_vs_expected_composites_tcga_precalculated.rds'))
obs <- info_all$obs_all
exp <- median(info_all$dat_all$prop)

## plot results (figure 1c in paper)
info_all$dat$group <- 'All samples'
plot_dat <- info_all$dat
plot_dat$data <- 'Randomly shuffled'
plot_dat$prop <- plot_dat$prop * 100
obs_dat <- data.table(group='All samples')
obs_dat$prop <- info_all$obs*100
obs_dat$data <- 'Observed'
pd <- rbind(plot_dat, obs_dat, fill=T)
x <- break_axis(pd$prop,maxlower=16,minupper=22,lowerticksize=1,upperticksize=1,ratio_lower_to_upper=0.5)
pd$prop2 <- x$newy
expected_color <- c('black','red')
names(expected_color) <- c('Randomly shuffled','Observed')
obs_label <- paste0(prettyNum(obs*100,digits=3),'%')

p <- ggplot(pd,aes(x=prop2)) +
    geom_histogram(binwidth=0.1,aes(fill=data),color='white',size=0.2) +    
    geom_vline(data=pd[data=='Observed',],aes(xintercept=prop2),color='red',linetype='dashed',size=0.2) +
    geom_text(data=pd[min(which(data=='Observed')),],x=17.5,label=obs_label, y=3000,color='red',size=5) +
    scale_fill_manual(values=expected_color,name=NULL) +
    scale_x_continuous(breaks=c(x$breaks), labels=c(x$labels), limits=c(13.0,19.1)) +
    scale_y_continuous(expand=c(0,0)) +
    labs(x='Percent of cases\n(TCGA pan-cancer, fixed gene content)',y='Frequency') + 
    theme_std(base_size=14) +
    theme(legend.position='none',axis.ticks.y=element_blank(),axis.text.y=element_blank())
p

```


***

\pagebreak

### EDF-1d

```{r panel_edf1d, width=6, height=5}

## extract/merge the shuffled composite rate
l <- readRDS(here('data/observed_vs_expected_composites_per_tmb_impact_precalculated.rds'))
burdens <- 1:50
extract_shuffled_data <- function(burden, l) {
    out <- data.table(mu=l[[burden]]$mu, lwr=l[[burden]]$lwr, upr=l[[burden]]$upr, burden=l[[burden]]$burden, obs=l[[burden]]$obs, N=l[[burden]]$samples,p=l[[burden]]$p)
    out
}
ldat <- lapply(burdens, extract_shuffled_data, l)
dat <- rbindlist(ldat)
dat$nlog10p <- -log10(dat$p)
dat$nlog10p[dat$nlog10p>5] <- 5

## fit 3rd order polynomial to points
m <- lm(mu ~ splines::bs(burden,3),data=dat)
dat$fit <- predict(m,newdata=dat,type='response',se.fit=F)
m.obs <- lm(obs ~ splines::bs(burden,3),data=dat)
dat$fit.obs <- predict(m.obs,newdata=dat,type='response',se.fit=F)

## compare the observed to the avg-shuffle per mutation burden
pval <- wilcox.test(dat$obs,dat$mu,alternative='two.sided',paired=T)$p.value
pval_lab <- paste0('p=',prettyNum(pval,digits=1))

plot_dat1 <- dat[,c('burden','fit'),with=F]
plot_dat1$type <- 'Expected'
setnames(plot_dat1,'fit','prop')
plot_dat1 <- plot_dat1[burden > 1,]

plot_dat2 <- dat[,c('burden','fit.obs'),with=F]
setnames(plot_dat2,'fit.obs','prop')
plot_dat2$type <- 'Observed'
plot_dat2 <- plot_dat2[burden > 1,]
plot_dat1$prop <- 100*plot_dat1$prop
plot_dat2$prop <- 100*plot_dat2$prop


## plot
cols <- c('black','#BFBEBD')
names(cols) <- c('Observed','Expected')

p <- ggplot(plot_dat1,aes(x=burden)) + 
    geom_line(data=plot_dat1,aes(y=prop,color=type),size = 0.75) +
    geom_line(data=plot_dat2,aes(y=prop,color=type),size = 0.75) +
    scale_x_continuous(breaks=seq(0,40,by=5),limits=c(0,41),expand=c(0,0)) +
    scale_y_continuous(limits=c(0,105),expand=c(0,0)) +
    scale_color_manual(values=cols,name=NULL) +
    geom_text(data=plot_dat1[1,],x=6,y=90,label=pval_lab,size=3.5) + 
    theme_std(base_size=16) + 
    labs(x='Mutational burden (non-syn/Mb)',y='% of cases with\ncomposite mutations') 
p

```



***

\pagebreak

# EDF-2

### EDF-2a

```{r panel_edf2a,width=8,height=10}
## number of higher-order composite overall
cols <- c(RColorBrewer::brewer.pal(8,'Paired'),'#BFBFBF')
cols <- cols[c(2,4,3,1,5:length(cols))]
names(cols) <- c('apobec','uv','smoking','mmr','pol','tmz','other','multiple','none')
cols[names(cols)=='other'] <- 'black'

check_singletons_vs_composites <- function(result) {
    num_variants <- nrow(result) 
    apobec <- sum(result$aetiology_class=='APOBEC') >= 1
    uv <- sum(result$aetiology_class=='UV') >= 1
    smoking <- sum(result$aetiology_class=='Smoking/Tobacco') >= 1
    decomp_failed <- sum(result$aetiology_class=='Decomposition failed') >= 1
    other <- sum(result$aetiology_class %in% 'Other') >= 1
    none <- !apobec & !uv & !smoking & !other & !decomp_failed
    list(num_variants=num_variants, 
         none=none,
         apobec=apobec, 
         uv=uv, 
         smoking=smoking, 
         other=other,
         decomp_failed=decomp_failed)
}

result <- fread(here('data/data_mutations_signature_attribution.txt.gz'))
result[aetiology_class %nin% c('No attributed signature','Decomposition failed','APOBEC','UV','Smoking/Tobacco'),aetiology_class:='Other']

## for singletons and composites, check if any are assigned to smoking/uv/apobec
x <- result[,check_singletons_vs_composites(.SD),by=c('metamaintype','Tumor_Sample_Barcode','Hugo_Symbol')]
x <- x[num_variants > 1,]
x <- melt(x,id.vars=c('metamaintype','Tumor_Sample_Barcode','Hugo_Symbol','num_variants'))
x <- x[value==T & variable!='decomp_failed',]
x$variable <- factor(x$variable, levels=c('none','apobec','uv','smoking','other'))

## merge numbers in composites with 6+ variants
xm <- xtabs(~num_variants + variable, data=x)
toadd <- colSums(xm[as.integer(rownames(xm)) >= 6,])
row_to_replace <- min(which(as.integer(rownames(xm))>=6))
xm[row_to_replace,] <- toadd
xm <- xm[1:row_to_replace,]
rownames(xm)[nrow(xm)] <- '6+'
cols <- cols[names(cols) %in% colnames(xm)]
num_variant_fields <- rownames(xm)
cnt_associated <- rowSums(xm[,names(cols)[names(cols)!='none']])
cnt_overall <- rowSums(xm)
pct_associated <- data.table(num_variants=rownames(xm),value=cnt_overall,prop_associated=paste0(round(100*(cnt_associated / cnt_overall),1),'%'))
xm <- xm/rowSums(xm)
xm <- adt(melt(xm))
xm$variable <- factor(xm$variable, levels=rev(names(cols)))
xm$num_variants <- factor(xm$num_variants, levels=rev(num_variant_fields))

## plot gene-level rate of composites associated with mutation signatures
p1 <- ggplot(xm,aes(x=num_variants,y=100*value)) +
    geom_bar(stat='identity',aes(fill=variable),width=0.8) +
    scale_fill_manual(values=cols, name='Associated signature') + 
    coord_flip() +
    scale_y_continuous(expand=c(0,0),position='right') + 
    theme_std(base_size=12) +
    theme(legend.position='bottom') +
    labs(y='% attributed to mutational signatures',
         x='Number of variants in composite mutation')
p1 <- extract_gglegend(p1)

t2 <- data.table(num_variants=names(cnt_overall), cnt_overall)
t2$num_variants <- factor(t2$num_variants,levels=rev(num_variant_fields))
t2$label <- t2$cnt_overall
t2$cnt_overall <- t2$cnt_overall / 1000
p2 <- ggplot(t2,aes(x=num_variants,y=cnt_overall)) + 
    geom_bar(stat='identity',width=0.8,fill='black') +
    geom_text(aes(label=label),hjust=-0.2,vjust=0.5) +
    scale_y_continuous(expand=c(0,0),position='right',limits=c(0,5),breaks=seq(0,4,by=1)) + 
    coord_flip() +
    theme_std(base_size=12) +
    theme( axis.ticks.y=element_blank(), axis.text.y=element_blank()) + #,axis.text.x=element_text(angle=-90,hjust=1,vjust=0.5)) +
    labs(x=NULL,y='N Composites\n(x1000)')

p <- plot_grid(p1$plot, p2, align='h', rel_widths=c(3,1))
p <- plot_grid(p,p1$legend, ncol=1,nrow=2, rel_heights=c(5,1))
p

```

***

\pagebreak

# EDF-2

### EDF-2b(i)

```{r panel_edf2bi, width=8, height=5} 
d <- fread(here('data/data_mutations_tcga.txt.gz')) 
clin <- fread(here('data/data_clinical_tcga.txt.gz'))
samples <- clin$Tumor_Sample_Barcode[clin$DISEASE=='SKCM']
d <- d[Tumor_Sample_Barcode %in% samples,]

## annotate mutations with if they are part of a composite
d$tsbgene <- paste(d$Tumor_Sample_Barcode,d$Hugo_Symbol)
tbl <- table.freq(d$tsbgene)
d <- merge(d, tbl, by.x='tsbgene', by.y='value', all.x=T)

## subset for SNPs, then merge with dyads centered within 1000bp of the SNP
d$start <- d$Start_Position-1000
d$end <- d$Start_Position+1000

## get mutation rate around active TFBSs
tabulate <- function(d, tfbs) {
    d <- d[Variant_Type=='SNP']
    setkey(d,'Chromosome','start','end')

    dd <- foverlaps(d, tfbs, type='any')
    dd_notfbs <- dd[is.na(tfbs_start),] ## mutations not within 1000bp of any tfbs
    dd$distance_to_tfbs <- dd$Start_Position-dd$tfbs_start
    dd[is.na(tfbs_start),distance_to_tfbs:=-9999]

    ## tabulate number of SNPs at any distance to tfbs
    x <- table.freq(dd$distance_to_tfbs)
    x$distance <- as.integer(x$value)
    x[,value:=NULL]
    x
}

active_tfbs_coding <- fread(here('data/tfbs_skcm_dhs_coding.txt'))
active_tfbs_coding <- active_tfbs_coding[,c('chr','tfbs_start','tfbs_end','tfbs_id'),with=F]
setkey(active_tfbs_coding,'chr','tfbs_start','tfbs_end')

dat <- tabulate(d, active_tfbs_coding)
dat$prop <- dat$N / sum(dat$N)
dat <- dat[order(distance),]
dat <- dat[distance >= -1e3 & distance <= 1e3]

## fit a smooth spline to the data
spl <- smooth.spline(x=dat$distance,y=dat$prop,spar=1)
pred <- predict(spl)
dat$pred <- pred$y

max_amp <- max(dat$pred)
dat$diff_from_halfmax <- abs(dat$pred - max_amp/2)
lwr <- which(dat$diff_from_halfmax==min(dat$diff_from_halfmax[dat$distance < 0]))
upr <- which(dat$diff_from_halfmax==min(dat$diff_from_halfmax[dat$distance > 0]))
FWHM <- dat$distance[c(lwr,upr)]
dm <- melt(dat[,c('distance','prop','pred'),with=F], id.vars='distance')
dm[variable=='prop',variable:='Observed']
dm[variable=='pred',variable:='Fit']

cols <- c('#BFBFBF','steelblue')
names(cols) <- c('Observed','Fit')

p0 <- ggplot(dm,aes(x=distance,y=value)) +
    geom_point(aes(color=variable)) +
    scale_color_manual(values=cols,name=NULL) + 
    geom_vline(xintercept=FWHM,color='black',linetype=2,size=0.5) +
    xlim(c(-1000,1000)) + ylim(c(0,7.5e-5)) +
    theme_std(base_size=16) + 
    theme(legend.position='none') +
    labs(x='Distance from active-TFBS [bp]',y='Proportion of SNVs')

## compare the fraction of singletons and composites to fall within the FWHM range
d_composite <- d[N > 1,]
d_singleton <- d[N == 1,]
ddc <- tabulate(d_composite, active_tfbs_coding)
ddc$class <- 'SNPs in composites'
ddc$prop <- ddc$N / sum(ddc$N)
dds <- tabulate(d_singleton, active_tfbs_coding)
dds$class <- 'SNPs in singletons'
dds$prop <- dds$N / sum(dds$N)
dd <- rbind(ddc,dds)
dd <- dd[order(distance,class)]
dd$within_fwhm <- dd$distance >= FWHM[1] & dd$distance <= FWHM[2]

x_c <- sum(dd$N[dd$within_fwhm==T & dd$class=='SNPs in composites'])
y_c <- sum(dd$N[dd$within_fwhm==F & dd$class=='SNPs in composites'])
N_c <- sum(dd$N[dd$class=='SNPs in composites'])
x_s <- sum(dd$N[dd$within_fwhm==T & dd$class=='SNPs in singletons'])
y_s <- sum(dd$N[dd$within_fwhm==F & dd$class=='SNPs in singletons'])
N_s <- sum(dd$N[dd$class=='SNPs in singletons'])

ci <- binom::binom.confint(c(x_s,x_c),c(N_s,N_c),method='exact')
ci$type <- c('Singleton','Composite')
ci$mean <- 100*ci$mean
ci$lower <- 100*ci$lower
ci$upper <- 100*ci$upper

m <- as.matrix(rbind(c(y_s,x_s),c(y_c,x_c)))
pval <- prop.test(m)$p.value
plabel <- paste0('P=',prettyNum(pval,digits=1))

p1 <- ggplot(ci,aes(x=type,y=mean)) +
    geom_bar(stat='identity',fill='black') +
    geom_signif(comparison=list(c('Composite','Singleton')),annotation=plabel,y_position=2) +
    scale_y_continuous(expand=c(0,0),limits=c(0,2.2)) + 
    geom_errorbar(aes(min=lower,max=mean,width=NA),size=1,color='white') +
    geom_errorbar(aes(min=mean,max=upper,width=NA),size=1,color='black') +
    theme_std(base_size=12) +
    theme(axis.text.x=element_text(angle=90,hjust=1,vjust=0.5)) +
    labs(x=NULL,y='Prop of SNVs proximal to\nactive coding TFBS')

p <- plot_grid(p0, p1, ncol=2, rel_widths=c(3,1.5))
p

```

***

\newpage

### EDF-2b(ii)

```{r panel_edf2bii, width=7, height=9} 

## check whether nucleosome dyad position correlates with composite mutations
## - for all mutations in composites, check if they are within a 2001bp window centered at a dyad
## - for the ones that are, get their distance from the nearest dyad, and then get the overall 
## - number of mutations at each distance from the nearest dyad.

d <- fread(here('data/data_mutations.txt.gz'), 
           select=c('metamaintype','exclude','putative_resistance_mutation','Tumor_Sample_Barcode',
                    'Hugo_Symbol','Chromosome','Start_Position','End_Position','Variant_Type'))

## annotate mutations with if they are part of a composite
d$tsbgene <- paste(d$Tumor_Sample_Barcode,d$Hugo_Symbol)
tbl <- table.freq(d$tsbgene)
d <- merge(d, tbl, by.x='tsbgene', by.y='value', all.x=T)

## subset for SNPs, then merge with dyads centered within 1000bp of the SNP
d$start <- d$Start_Position-1000
d$end <- d$Start_Position+1000
d <- d[exclude==F,] 

dyads <- fread(here('data/dyads_genic.bed'))
names(dyads) <- c('Chromosome','start','end')
dyads$Chromosome <- gsub('chr','',dyads$Chromosome)
setkey(dyads,'Chromosome','start','end')

tabulate <- function(d, dyads) {
    d <- d[Variant_Type=='SNP']
    setkey(d,'Chromosome','start','end')

    dd <- foverlaps(d, dyads, type='any')
    dd_nodyad <- dd[is.na(start),] ## mutations not within 1000bp of any dyad
    dd$distance_to_dyad <- dd$Start_Position-dd$start
    dd[is.na(start),distance_to_dyad:=-9999]

    ## tabulate number of SNPs at any distance to dyad
    x <- table.freq(dd$distance_to_dyad)
    x$distance <- as.integer(x$value)
    x[,value:=NULL]
    x
}

dat <- tabulate(d, dyads)
dat$prop <- dat$N / sum(dat$N)
dat <- dat[order(distance),]
dat <- dat[distance >= -1e3 & distance <= 1e3]

## fit a smooth spline to the data
spl <- smooth.spline(x=dat$distance,y=dat$prop,spar=1)
pred <- predict(spl)
dat$pred <- pred$y

max_amp <- max(dat$pred)
dat$diff_from_halfmax <- abs(dat$pred - max_amp/2)
lwr <- which(dat$diff_from_halfmax==min(dat$diff_from_halfmax[dat$distance < 0]))
upr <- which(dat$diff_from_halfmax==min(dat$diff_from_halfmax[dat$distance > 0]))
FWHM <- dat$distance[c(lwr,upr)]
dm <- melt(dat[,c('distance','prop','pred'),with=F], id.vars='distance')
dm[variable=='prop',variable:='Observed']
dm[variable=='pred',variable:='Fit']

cols <- c('#BFBFBF','steelblue')
names(cols) <- c('Observed','Fit')

p0 <- ggplot(dm,aes(x=distance,y=value)) +
    geom_point(aes(color=variable)) +
    scale_color_manual(values=cols,name=NULL) + 
    geom_vline(xintercept=FWHM,color='black',linetype=2,size=0.5) +
    xlim(c(-1000,1000)) + ylim(c(0,0.0003)) +
    theme_std(base_size=16) + 
    theme(legend.position='none') +
    labs(x='Distance from dyad [bp]',y='Proportion of SNVs')

## compare the fraction of singletons and composites to fall within the FWHM range
d_composite <- d[N > 1,]
d_singleton <- d[N == 1,]
ddc <- tabulate(d_composite, dyads)
ddc$class <- 'SNPs in composites'
ddc$prop <- ddc$N / sum(ddc$N)
dds <- tabulate(d_singleton, dyads)
dds$class <- 'SNPs in singletons'
dds$prop <- dds$N / sum(dds$N)
dd <- rbind(ddc,dds)
dd <- dd[order(distance,class)]
dd$within_fwhm <- dd$distance >= FWHM[1] & dd$distance <= FWHM[2]

x_c <- sum(dd$N[dd$within_fwhm==T & dd$class=='SNPs in composites'])
y_c <- sum(dd$N[dd$within_fwhm==F & dd$class=='SNPs in composites'])
N_c <- sum(dd$N[dd$class=='SNPs in composites'])
x_s <- sum(dd$N[dd$within_fwhm==T & dd$class=='SNPs in singletons'])
y_s <- sum(dd$N[dd$within_fwhm==F & dd$class=='SNPs in singletons'])
N_s <- sum(dd$N[dd$class=='SNPs in singletons'])

ci <- binom::binom.confint(c(x_s,x_c),c(N_s,N_c),method='exact')
ci$type <- c('Singleton','Composite')
ci$mean <- 100*ci$mean
ci$lower <- 100*ci$lower
ci$upper <- 100*ci$upper

m <- as.matrix(rbind(c(y_s,x_s),c(y_c,x_c)))
pval <- prop.test(m)$p.value
plabel <- paste0('P=',prettyNum(pval,digits=1))

p1 <- ggplot(ci,aes(x=type,y=mean)) +
    geom_bar(stat='identity',fill='black') +
    geom_signif(comparison=list(c('Composite','Singleton')),annotation=plabel,y_position=11.5) +
    scale_y_continuous(expand=c(0,0),limits=c(0,12.5)) + 
    geom_errorbar(aes(min=lower,max=mean,width=NA),size=1,color='white') +
    geom_errorbar(aes(min=mean,max=upper,width=NA),size=1,color='black') +
    theme_std(base_size=12) +
    theme(axis.text.x=element_text(angle=90,hjust=1,vjust=0.5)) +
    labs(x=NULL,y='Prop of SNVs proximal to\nnucleosome dyad')


p <- plot_grid(p0, p1, ncol=2, rel_widths=c(3,1.5))
p
```

***

\pagebreak

# EDF-3

### EFD-3a

```{r panel_edf3a,width=8,height=6}
d <- fread(here('data/data_mutations.txt.gz'), select=c('metamaintype','Tumor_Sample_Barcode','Hugo_Symbol','exclude','putative_resistance_mutation','Role'))
d <- d[putative_resistance_mutation==F & exclude==F, c('Role','Tumor_Sample_Barcode','Hugo_Symbol'),with=F]

count_patients <- function(d) {
    tbl <- table.freq(d$Tumor_Sample_Barcode)
    samples <- tbl$value
    composite_samples <- tbl$value[tbl$N > 1]
    nc <- length(unique(strtrim(composite_samples,9)))
    singleton_samples <- tbl$value[tbl$N == 1]
    ns <- length(unique(strtrim(singleton_samples,9)))
    list(n_singleton=ns, n_composite=nc)
}
info <- d[,count_patients(.SD),by=c('Hugo_Symbol','Role')]
info$n_patients <- info$n_singleton + info$n_composite
info[Role=='Oncogene/TSG',Role:='Other']
info[Role=='n/a',Role:='Other']
info[is.na(Role),Role:='Other']
genes <- fread(here('data/gene_enrichment_precalculated.txt'))
genes <- genes$Hugo_Symbol[genes$q_enriched < 0.01]
info <- info[Hugo_Symbol %in% genes,]
subset <- function(info, role, n=10) {
    info <- info[Role %in% role,]
    if(n > nrow(info)) n <- nrow(info)
    info <- info[order(n_patients,decreasing=T),]
    info <- info[1:n,]
    info$i <- nrow(info):1
    info
}
dat1 <- subset(info,'TSG',n=10)
dat2 <- subset(info,'Other',n=10)
dat3 <- subset(info,'Oncogene',n=10)

dat <- rbind(dat1, dat2, dat3)
dat$gene <- paste(dat$Role,dat$Hugo_Symbol)
dat$Role <- factor(dat$Role, levels=c('Other','TSG','Oncogene'))

dat <- dat[order(Role,n_composite, decreasing=F),]
dat$pct_composite <- round(100*dat$n_composite/dat$n_patients,1)
dat$gene <- factor(dat$gene, levels=dat$gene)
dat[,c('i'):=NULL]
dm <- melt(dat[,c('Hugo_Symbol','Role','n_composite','n_patients','pct_composite','gene'),with=F],id.vars=c('Hugo_Symbol','Role','gene','pct_composite'))
dm$variable <- gsub('n_','',dm$variable)
dm[variable=='patients',label:=value]
dm[variable=='patients',value:=as.integer(-1*value)]
dm[Hugo_Symbol!='TP53',label:=NA]
dm$Hugo_Symbol <- factor(dm$Hugo_Symbol, levels=unique(dat$Hugo_Symbol))
dm$value <- as.numeric(dm$value)
dm[value < 0,value:=as.numeric(value/1000)]
dm[value > 0,value:=as.numeric(value/100)]
dm[Hugo_Symbol=='TP53' & variable=='patients',value:=-8.75]

cols <- c('#EC7C65','#1E88AC')
names(cols) <- c('composite','patients')

p1 <- ggplot(dm,aes(x=Hugo_Symbol,y=value)) +
    geom_bar(stat='identity',aes(fill=variable)) +
    geom_text(color='white',aes(label=label),hjust=-1.25,size=5) +
    scale_fill_manual(values=cols,name=NULL) + 
    scale_y_continuous(expand=c(0,0),limits=c(-10,10),
                       breaks=c(seq(-8,0,by=2),seq(0,8,by=2)),labels=c(seq(8,0,by=-2),seq(0,800,by=200)),position='right') +
    theme_std(base_size=12) + 
    theme(axis.line.y=element_blank(),axis.ticks.y=element_blank(),legend.position='none') +
    coord_flip() +
    labs(x=NULL,y='N mutant patients (x1000)        Composite cases')

dat$Hugo_Symbol <- factor(dat$Hugo_Symbol, levels=levels(dm$Hugo_Symbol))
dat$label <- dat$pct_composite
dat[Hugo_Symbol!='APC',label:=NA]
dat[Hugo_Symbol=='APC',pct_composite:=23]

p2 <- ggplot(dat,aes(x=Hugo_Symbol,y=pct_composite)) +
    geom_bar(stat='identity') +
    geom_text(color='white',aes(label=label),hjust=2,size=5) +
    scale_y_continuous(expand=c(0,0),limits=c(0,25),
                       breaks=seq(0,20,by=5),position='right') + 
    theme_std(base_size=12) + 
    theme(axis.line.y=element_blank(),axis.ticks.y=element_blank(),legend.position='none',
          axis.text.y=element_blank()) +
    coord_flip() +
    labs(x=NULL,y='% composite mutant')

p <- plot_grid(p1,p2,rel_widths=c(3,1),align='h',ncol=2,nrow=1)
p
```

***

\pagebreak

### EFD-3b

```{r panel_edf3b, width=8, height=5} 

## QQ plot entirely for oncogenes

gene_info <- fread(here('data/gene_info.txt'),
                   select=c('Hugo_Symbol','Role','panel_introduced','reptime','percentage_gene_gc_content',
                            'cds_length','hic'))

prep_data <- function(d) {
    d <- d[exclude==F & 
           putative_resistance_mutation==F,c('Tumor_Sample_Barcode','Hugo_Symbol','Variant_Type','Variant_Classification','tcn',
                                             'putative_resistance_mutation','mutationid','high_tmb','hotspot_class'),with=F]
    d <- merge(d, gene_info, by='Hugo_Symbol', all.x=T)

    ## consider the TERT promoter as a unique gene, where the cds_length is now the total length of the promoter region covered by IMPACT
    tert_promoter <- which(d$Hugo_Symbol=='TERT' & d$Variant_Classification=='TERT promoter')
    d[tert_promoter, Hugo_Symbol:='TERT promoter']
    d[tert_promoter, cds_length:=499] ## this is the length of the promoter region of TERT covered by IMPACT468
    d[tert_promoter, percentage_gene_gc_content:=0.7896] ## this is the %GC content of TERT promoter region
    
    get_composite_rate_per_gene <- function(d) {
        tbl <- table.freq(d$Tumor_Sample_Barcode)
        samples <- tbl$value
        composite_samples <- tbl$value[tbl$N > 1]
        singleton_samples <- tbl$value[tbl$N == 1]

        ## don't overcount multiple samples per patient
        samples <- length(unique(strtrim(samples, 9)))
        composites <- length(unique(strtrim(composite_samples, 9)))
        singletons <- length(unique(strtrim(singleton_samples, 9)))
        tcn <- mean(d$tcn,na.rm=T)
        list(singletons=singletons,composites=composites,samples=samples,tcn=tcn,
             reptime=unique(d$reptime),percentage_gene_gc_content=unique(d$percentage_gene_gc_content),cds_length=unique(d$cds_length),
             hic=unique(d$hic),
             panel_introduced=unique(d$panel_introduced),Role=unique(d$Role))
    }

    res <- d[,get_composite_rate_per_gene(.SD),by=Hugo_Symbol]
    res[is.nan(tcn),tcn:=2]
    res$hic[is.na(res$hic)] <- round(mean(res$hic,na.rm=T))
    res$panel_introduced <- factor(res$panel_introduced)
    res$Hugo_Symbol <- gsub('-','_',res$Hugo_Symbol)
    res$Hugo_Symbol <- gsub(' ','_',res$Hugo_Symbol)
    res
}

d <- fread(here('data/data_mutations.txt.gz'))
d_onc <- prep_data(d[Role=='Oncogene',])
d_onc$class <- 'Oncogenes'
#d_onc <- merge(d_onc, obs, by='Hugo_Symbol', all.x=T)
d_onc$observed_proportion <- d_onc$composites / d_onc$samples

# Run the gene-enrichment binomial test on the different prepped datasets
test_data <- function(res) { 
    m <- glm.nb(composites ~ offset(log(samples)) + reptime + cds_length + percentage_gene_gc_content + panel_introduced + tcn + hic, data=res)
    summary(m)$coefficients
    predictions <- predict(m,newdata=res,type='response')
    res$predicted_proportion <- predictions / res$samples
    dat <- res[,c('Hugo_Symbol','composites','samples','predicted_proportion','observed_proportion','composites','samples','Role'),with=F]

    test_gene <-function(dat) {
        x <- tryCatch({
            tst <- binom.test(dat$composites,dat$samples,dat$predicted_proportion,alternative='greater')
            p.value <- tst$p.value
            list(p.value=p.value)
        },error=function(e) {
            list(p.value=as.numeric(NA))
        })
        dat$p.value <- x$p.value
        dat
    }
    dat <- dat[,test_gene(.SD),by=Hugo_Symbol]
    dat$logOR <- log2(dat$observed_proportion/dat$predicted_proportion)
    dat <- dat[order(p.value,decreasing=F),]
    dat$q.value <- p.adjust(dat$p.value,method='BH')
    dat
}

result_onc <- test_data(d_onc)
result_onc <- result_onc[,c('Hugo_Symbol','p.value','q.value'),with=F]
names(result_onc) <- c('Hugo_Symbol','p_enriched','q_enriched')

# generate qqplot
get_qqplot <- function(dat) {
    N = nrow(dat)
    ci <- 0.95
    df <- data.table(observed=-log10(dat$p_enriched),
                     expected=-log10(1:N/N),
                     p=dat$p_enriched,
                     q=dat$q_enriched,
                     gene=dat$Hugo_Symbol,
                     cupper=-log10(qbeta(ci,1:N, N - 1:N + 1)),
                     clower=-log10(qbeta(1 - ci, 1:N, N - 1:N + 1)))
    df$Significance <- ifelse(df$q < 0.01,'q < 0.01','Not significant')
    df$Significance <- factor(df$Significance,levels=c('q < 0.01','Not significant'))
    df$label <- df$gene
    df$label[df$q >= 0.01] <- NA

    ## make plot
    log10Pe = expression(paste("Expected -log"[10], plain(P)))
    log10Po = expression(paste("Observed -log"[10], plain(P)))
    mycols <- c('#a6a6a6','#e50000')
    names(mycols) <- c('Not significant','q < 0.01')
    p <- ggplot(df,aes(x=expected,y=observed)) +
        geom_point(pch=19, size=2.5, aes(color=Significance)) +
        geom_abline(intercept=0, slope=1, alpha=0.5) +
        geom_line(aes(expected, cupper), linetype=2) +
        geom_line(aes(expected, clower), linetype=2) +
        scale_color_manual(values=mycols) +
        xlab(log10Pe) +
        ylab(log10Po) +
        geom_text_repel(aes(label=label), color='black',size=2.5, na.rm=T) +
        theme_std(base_size=16)
    p
}

p1 <- get_qqplot(result_onc) + 
    ylim(80,100) +
    theme(legend.position='none',axis.line.x=element_blank(),axis.ticks.x=element_blank(),axis.text.x=element_blank()) +
    labs(x=NULL,y=NULL,subtitle='Enrichment for composite mutations among oncogenes') 
    
p2 <- get_qqplot(result_onc) + 
    ylim(0,10) +
    theme(legend.position='none')

p <- plot_grid(p1,p2,rel_heights=c(1,3),align='v',ncol=1,nrow=2)
p
```


***

\pagebreak

# EDF-4

```{r panel_edf4, width=8, height=5} 

## define helper functions here
extract_cis_trans <- function(d) {
    ## don't overcount multiple pairs of composite mutations in the same sample+gene+phase
    d$id <- paste(d$Tumor_Sample_Barcode,d$Hugo_Symbol,d$phase)

    ## don't overcount 3+ mutation composite-mutations; using unique sample/gene/Class counts
    d <- d[!duplicated(id),c('Tumor_Sample_Barcode','metamaintype','Hugo_Symbol','Role','phase','id',
                             'putative_resistance_mutation.1','putative_resistance_mutation.2',
                             'HGVSp_Short.1','HGVSp_Short.2',
                               'ccf_Mcopies.1','clonal.1','ccf_Mcopies.2','clonal.2','important.1','important.2',
                               'common_reads_alt1_alt2','common_reads_alt1_ref2','common_reads_ref1_alt2','common_reads_ref1_ref2',
                               't_depth.1','t_depth.2'),with=F]
    d
}

label_composite <- function(d) {
    label <- ''
    if(d$Hugo_Symbol=='EGFR' & (d$HGVSp_Short.1=='Exon 19 INDELs' & d$HGVSp_Short.2=='T790M'))  label <- 'EGFR Exon 19 INDELs + T790M'
    if(d$Hugo_Symbol=='EGFR' & (d$HGVSp_Short.1=='T790M' & d$HGVSp_Short.2=='L858R'))  label <- 'EGFR L858R + T790M'
    if(d$Hugo_Symbol=='EGFR' & (d$HGVSp_Short.1=='D761Y' & d$HGVSp_Short.2=='T790M'))  label <- 'EGFR D761Y + T790M'
    if(d$Hugo_Symbol=='EGFR' & (d$HGVSp_Short.1=='G719A' & d$HGVSp_Short.2=='T790M'))  label <- 'EGFR G719A + T790M'
    if(d$Hugo_Symbol=='EGFR' & (d$HGVSp_Short.1=='T790M' & d$HGVSp_Short.2=='C797S'))  label <- 'EGFR T790M + C797S'
    if(d$Hugo_Symbol=='EGFR' & (d$HGVSp_Short.1=='Exon 19 INDELs' & d$HGVSp_Short.2=='C797S'))  label <- 'EGFR Exon 19 INDELs + C797S'
    if(d$Hugo_Symbol=='EGFR' & (d$HGVSp_Short.1=='C797S' & d$HGVSp_Short.2=='L858R'))  label <- 'EGFR L858R + C797S'
    if(d$Hugo_Symbol=='EGFR' & (d$HGVSp_Short.1=='D761Y' & d$HGVSp_Short.2=='C797S'))  label <- 'EGFR D761Y + C797S'
    if(d$Hugo_Symbol=='EGFR' & (d$HGVSp_Short.1=='G719A' & d$HGVSp_Short.2=='C797S'))  label <- 'EGFR G719A + C797S'
    #if(label=='' & d$Hugo_Symbol=='EGFR' & (d$HGVSp_Short.1=='T790M'  | d$HGVSp_Short.2=='T790M'))  label <- 'EGFR Other driver + T790M'
    #if(label=='' & d$Hugo_Symbol=='EGFR' & (d$HGVSp_Short.1=='C797S'  | d$HGVSp_Short.2=='C797S'))  label <- 'EGFR Other driver + C797S'
    if(d$Hugo_Symbol=='AR' &   (d$HGVSp_Short.1=='L702H'  & d$HGVSp_Short.2=='H875Y'))  label <- 'AR L702H + H875Y'
    if(d$Hugo_Symbol=='ALK' &  (d$HGVSp_Short.1=='D1203N' & d$HGVSp_Short.2=='L1196M')) label <- 'ALK D1203N + L1196M'
    if(d$Hugo_Symbol=='ALK' &  (d$HGVSp_Short.1=='G1202R' & d$HGVSp_Short.2=='L1196M')) label <- 'ALK G1202R + L1196M'
    if(d$Hugo_Symbol=='BRCA1' &  (d$HGVSp_Short.1=='E1257Gfs*9' & d$HGVSp_Short.2=='N1235Tfs*18')) label <- 'BRCA1 Frame-restoring INDELs'

    if(label=='')  label <- paste(d$Hugo_Symbol,d$HGVSp_Short.1,'+',d$HGVSp_Short.2)
    d$label <- label
    d
}

summarize_composite_with_alleles <- function(d) {
    n <- length(unique(strtrim(d$Tumor_Sample_Barcode,9)))
    list(n=n)
}


## load phased data, add phasing by LOH with FACETS
d <- fread(here('data/data_mutations_phased.txt.gz'))
d <- d[exclude==F & (putative_resistance_mutation.1==T | putative_resistance_mutation.2==T),]
reviewed <- fread(here('data/resistance_mutation_phasing_reviewed.txt'))
reviewed <- reviewed[,c('Tumor_Sample_Barcode','Hugo_Symbol','HGVSp_Short.1','HGVSp_Short.2','phase_reviewed'),with=F]
d <- merge(d, reviewed[,c('Tumor_Sample_Barcode','Hugo_Symbol','HGVSp_Short.1','HGVSp_Short.2','phase_reviewed'),with=F],
           by=c('Tumor_Sample_Barcode','Hugo_Symbol','HGVSp_Short.1','HGVSp_Short.2'),all.x=T)
d[phase_reviewed %in% c('cis','cis*'),phase:=phase_reviewed]
to_remove <- c(
               ## bad facets fit in this sample, we can't be sure variants are in the same cell: exclude 
               which(d$Tumor_Sample_Barcode=='P-0013110-T01-IM5' & d$tm.1=='KIT C809' & d$tm.2=='KIT indel:820-821'),
               ## this sample has EGFR S768 cis with EGFR T790, but both are in trans with a passenger C775, so I am removing the passenger composite
               which(d$Tumor_Sample_Barcode=='P-0003923-T01-IM5' & d$tm.1=='EGFR C775' & d$tm.2=='EGFR T790'), 
               which(d$Tumor_Sample_Barcode=='P-0003923-T01-IM5' & d$tm.1=='EGFR S768' & d$tm.2=='EGFR C775'))
d <- d[-to_remove,]

## get barplot of resistance mutations per gene
d$important.1 <- d$hotspot.1==T | d$oncogenic.1 %in% c('Oncogenic','Likely Oncogenic','Predicted Oncogenic') | 
    (d$Role %in% c('TSG','Oncogene/TSG') & d$truncating.1==T)
d$important.2 <- d$hotspot.2==T | d$oncogenic.2 %in% c('Oncogenic','Likely Oncogenic','Predicted Oncogenic') | 
    (d$Role %in% c('TSG','Oncogene/TSG') & d$truncating.2==T)

d$HGVSp_Short.1 <- gsub('p[.]','',d$HGVSp_Short.1)
d$HGVSp_Short.2 <- gsub('p[.]','',d$HGVSp_Short.2)
d[Variant_Classification.1 %in% c('In_Frame_Ins','In_Frame_Del') & Hugo_Symbol=='EGFR' & Exon_Number.1=='19/28',HGVSp_Short.1:=paste('Exon 19 INDELs')]
d[Variant_Classification.1 %in% c('In_Frame_Ins','In_Frame_Del') & Hugo_Symbol=='EGFR' & Exon_Number.1=='20/28',HGVSp_Short.1:=paste('Exon 20 INDELs')]
d[Variant_Classification.2 %in% c('In_Frame_Ins','In_Frame_Del') & Hugo_Symbol=='EGFR' & Exon_Number.2=='19/28',HGVSp_Short.2:=paste('Exon 19 INDELs')]
d[Variant_Classification.2 %in% c('In_Frame_Ins','In_Frame_Del') & Hugo_Symbol=='EGFR' & Exon_Number.2=='20/28',HGVSp_Short.2:=paste('Exon 20 INDELs')]
d[Variant_Classification.1 %in% c('In_Frame_Ins','In_Frame_Del') & Hugo_Symbol=='KIT' & Exon_Number.1=='11/21',HGVSp_Short.1:=paste('Exon 11 INDELs')]
d[Variant_Classification.2 %in% c('In_Frame_Ins','In_Frame_Del') & Hugo_Symbol=='KIT' & Exon_Number.2=='11/21',HGVSp_Short.2:=paste('Exon 11 INDELs')]
d <- d[phase %in% c('cis','cis*','trans'),]
d[phase=='cis*',phase:='cis']
d$id <- paste(strtrim(d$Tumor_Sample_Barcode,9),d$Hugo_Symbol,d$HGVSp_Short.1,d$HGVSp_Short.2,d$phase)
d <- d[!duplicated(id),]

## prep data for plot
dd <- extract_cis_trans(d)
dd <- dd[order(Tumor_Sample_Barcode,Hugo_Symbol),]
dd$i <- 1:nrow(dd)
dd <- dd[,label_composite(.SD),by=i]
toremove <- which(dd$Hugo_Symbol=='EGFR' & (grepl('T790M',dd$label)==F & grepl('C797S',dd$label)==F))
dd <- dd[-toremove,]
dd <- dd[Hugo_Symbol!='ESR1',]
dd[label=='BRCA1 S1253Rfs*10 + N1235Tfs*18',label:='BRCA1 frame-restoring']
info <- dd[,summarize_composite_with_alleles(.SD),by=c('Hugo_Symbol','label','phase')]

info <- info[phase %in% c('cis','trans'),]
summarize_gene <- function(info) list(n=sum(info$n)) 
ns <- info[,summarize_gene(.SD),by=Hugo_Symbol]
ns <- ns[order(n,decreasing=T),]
info$Hugo_Symbol <- factor(info$Hugo_Symbol, levels=ns$Hugo_Symbol)
info <- info[order(info$Hugo_Symbol,info$n,decreasing=F),]
info$label <- factor(info$label, levels=info$label)

## color scheme
gene_colors <- c(
                 (brewer.pal(9,'Purples')[c(4,7)]), ## EGFR
                 (brewer.pal(9,'Blues')[c(2:9)]), ## KIT
                 (brewer.pal(9,'Greens')[c(3,5,7)]), ## AR
                 (brewer.pal(9,'Reds')[c(3,7)]), ## ALK
                 'pink') ## brca1
names(gene_colors) <- info$label

## barplot for resistance mutations by gene
p <- ggplot(info, aes(x=Hugo_Symbol, y=n)) +
    geom_bar(stat='identity',aes(fill=label)) +
    theme_std(base_size=16) +
    scale_y_continuous(expand=c(0,0),breaks=seq(0,14,by=2),limits=c(0,14)) + 
    scale_fill_manual(values=gene_colors, name='Resistance mutations') +
    labs(x=NULL,y='N in cis resistance mutations') +
    theme(axis.text.x=element_text(angle=90,hjust=1,vjust=0.5)) 
p
```

***

\pagebreak 

# EDF-7

### EDF-7a

```{r panel_edf7a}

# for combined composites across all genes, get the
# breakdown of associated signatures
result <- fread(here('data/data_mutations_signature_attribution.txt.gz'))
result$aetiology <- factor(result$aetiology, 
                           levels=c('Decomposition failed','Other','Multiple, non-separable','TMZ',
                                    'POLE','MMR','Smoking/Tobacco','UV','APOBEC','No attributed signature'))

# count number of variants per sample/gene
count_variants <- function(result) {
    num_variants <- nrow(result)
    tbl <- table(result$aetiology)
    tbl <- c(num_variants, tbl)
    names(tbl)[1] <- 'num_variants'
    tbl <- as.list(tbl)
    tbl
}
x <- result[,count_variants(.SD),by=c('metamaintype','Tumor_Sample_Barcode','Hugo_Symbol')]
x <- x[num_variants > 1,]
x <- melt(x,id.vars=c('metamaintype','Tumor_Sample_Barcode','Hugo_Symbol','num_variants'))
x <- x[variable!='Decomposition failed',]
x <- x[value > 0,] ## these are composites where 1+ variants in a sample/gene are attributed to the given aetiology
x[variable %nin% c('No attributed signature','APOBEC','UV','Smoking/Tobacco'),variable:='Other']
x$variable <- as.character(x$variable)
x[variable=='No attributed signature',variable:='Aging or no attributed mutational signature']
x[variable=='Smoking/Tobacco',variable:='Smoking']

cols <- c('#CCCCCC','#1E78B4','#36A048','#B4D789','black')
names(cols) <- c('Aging or no attributed mutational signature','APOBEC','UV','Smoking','Other')
x$variable <- factor(x$variable, levels=names(cols))
tbl <- table.freq(x$variable)
names(tbl) <- c('aetiology','N')

coldata <- data.table(col=cols, aetiology=names(cols))
tbl <- merge(tbl, coldata, by='aetiology', all.x=T)
tbl$aetiology <- paste0(tbl$aetiology,' (',tbl$N,')')
tbl$prop <- 100*(tbl$N / sum(tbl$N))
tbl <- tbl[order(prop, decreasing=T),]
tbl$aetiology <- factor(tbl$aetiology, levels=rev(tbl$aetiology))
tbl$ypos <- cumsum(tbl$prop)
tmpcols <- tbl$col
names(tmpcols) <- tbl$aetiology
N <- sum(tbl$N)

p <- ggplot(tbl,aes(x=1,y=prop)) +
    geom_bar(stat='identity',aes(fill=aetiology)) +
    scale_fill_manual(values=tmpcols, name=paste0('Associated\nsignature (N=',N,')')) +
    theme_std(base_size=16) +
    scale_y_continuous(expand=c(0,0),limits=c(0,100),breaks=seq(0,100,by=25)) +
    theme(axis.ticks.x=element_blank(),axis.text.x=element_blank(),legend.position='right') +
    xlim(0,2) +
    labs(x=NULL,y='Number of all composite mutations')
p

```

***

\pagebreak

### EDF-7b

```{r panel_edf7b,width=7,height=8}

## show freq of variants with attributal signature per gene
result <- fread(here('data/data_mutations_signature_attribution.txt.gz'))
result <- result[aetiology_class!='Decomposition failed',]

get_attributed_signatures <- function(result) {
    num_variants <- nrow(result)
    apobec <- sum(result$aetiology_class=='APOBEC') >= 1
    uv <- sum(result$aetiology_class=='UV') >= 1
    smoking <- sum(result$aetiology_class=='Smoking/Tobacco') >= 1
    multiple <- sum(result$aetiology_class=='Multiple, non-separable') >= 1
    mmr <- sum(result$aetiology_class=='MMR') >= 1
    other <- sum(result$aetiology_class=='Other') >= 1
    pol <- sum(result$aetiology_class=='POLE') >= 1
    tmz <- sum(result$aetiology_class=='TMZ') >= 1
    none <- !apobec & !uv & !smoking & !multiple & !mmr & !other & !pol & !tmz
    list(num_variants=num_variants,
         none=none,
         apobec=apobec,
         uv=uv,
         smoking=smoking,
         multiple=multiple,
         mmr=mmr,
         other=other,
         pol=pol,
         tmz=tmz)
}

## for singletons and composites, check if any are assigned to smoking/uv/apobec
x <- result[,get_attributed_signatures(.SD),by=c('metamaintype','Tumor_Sample_Barcode','Hugo_Symbol')]
x <- x[num_variants > 1,]
x <- melt(x,id.vars=c('metamaintype','Tumor_Sample_Barcode','Hugo_Symbol','num_variants'))
x <- x[value==T,]
x$variable <- as.character(x$variable)

x[variable=='apobec',variable:='APOBEC']
x[variable=='uv',variable:='UV']
x[variable=='smoking',variable:='Smoking/Tobacco']
x[variable=='mmr',variable:='MMR']
x[variable=='pol',variable:='POLE']
x[variable=='tmz',variable:='TMZ']
x[variable=='other',variable:='Other']
x[variable=='multiple',variable:='Multiple, non-separable']
x[variable=='none',variable:='No attributed signature']

x[variable %nin% c('No attributed signature','APOBEC','UV','Smoking/Tobacco'),variable:='Other']
x[variable=='No attributed signature',variable:='Aging or no attributed mutational signature']
x[variable=='Smoking/Tobacco',variable:='Smoking']

cols <- c('#CCCCCC','#1E78B4','#36A048','#B4D789','black')
names(cols) <- c('Aging or no attributed mutational signature','APOBEC','UV','Smoking','Other')
x$variable <- factor(x$variable, levels=rev(names(cols)))

## subset for recurrently composite-mutated genes
tbl <- table.freq(x$Hugo_Symbol)
gene_enrichment2 <- fread(here('data/gene_enrichment_precalculated.txt'))
recurrent_genes <- gene_enrichment2$Hugo_Symbol[gene_enrichment2$q_enriched < 0.01]
x <- x[Hugo_Symbol %in% recurrent_genes,]
tbl <- table.freq(x$Hugo_Symbol)

xm <- as.data.frame.matrix(xtabs(~Hugo_Symbol + variable,data=x))
xm <- cbind(Hugo_Symbol=rownames(xm), adt(xm))
cnt_associated <- rowSums(xm[,2:5])
cnt_overall <- rowSums(xm[,2:ncol(xm),with=F])
pct_associated <- data.table(Hugo_Symbol=xm$Hugo_Symbol,value=cnt_overall,prop_associated=paste0(round(100*(cnt_associated / cnt_overall),1)))
xm <- adt(melt(xm,id.vars='Hugo_Symbol'))
xm$variable <- factor(xm$variable, levels=(names(cols)))
xm$Hugo_Symbol <- factor(xm$Hugo_Symbol, levels=(tbl$value))

## plot gene-level rate of composites associated with mutation signatures
p <- ggplot(xm,aes(x=Hugo_Symbol,y=value)) +
    geom_bar(stat='identity',aes(fill=variable),width=0.8) +
    scale_fill_manual(values=cols, name='Associated signature') +
    scale_y_continuous(expand=c(0,0)) +
    theme_std(base_size=16) +
    theme(axis.text.x=element_text(angle=90,hjust=1,vjust=0.5)) +
    labs(x='Genes enriched for composite mutations',y='N composite mutations')
p <- extract_gglegend(p)
p <- plot_grid(p$plot, p$legend, ncol=1, nrow=2, rel_heights=c(3,2))
p
```

***

\pagebreak

# EDF-8

### EDF-8b(i)

```{r, panel_edf8bi, fig.width=6,fig.height=4}


d_clinical <- fread(here('data/data_clinical.txt.gz'),select=c('Tumor_Sample_Barcode','exclude'))

## load starting datasets
d_mutations <- fread(here('data/data_mutations.txt.gz'),select=c('Tumor_Sample_Barcode','PATIENT_ID','exclude','Hugo_Symbol','Role','putative_resistance_mutation',
                                                              'Variant_Classification','Variant_Type','tcn', 'mutationid', 'high_tmb','hotspot_class',
                                                              'tm','hotspotid','HGVSp_Short','truncating','Reference_Allele','Tumor_Seq_Allele2',
                                                              'Start_Position','End_Position'))

tert <- d_mutations
tert <- tert[Hugo_Symbol=='TERT' & Variant_Classification=='TERT promoter' & exclude==F,]
get_tert_mutation <- function(Start_Position,Reference_Allele,Tumor_Seq_Allele2) {
    Amino_Acid_Position <- as.integer(substr(Start_Position,5,7))
    Reference_Allele <- toupper(Reference_Allele)
    Tumor_Seq_Allele2 <- toupper(Tumor_Seq_Allele2)
    mutation <- paste0(Amino_Acid_Position,Reference_Allele,'>',Tumor_Seq_Allele2)
    mutation
}
tert$mutation <- get_tert_mutation(tert$Start_Position, tert$Reference_Allele, tert$Tumor_Seq_Allele2)
mutations <- unique(tert$mutation)

getN <- function(query.mutation, d) {
    mutant_samples <- d$Tumor_Sample_Barcode[d$mutation==query.mutation]
    samples_with_mutation <- d$Tumor_Sample_Barcode[d$Tumor_Sample_Barcode %in% mutant_samples]
    tbl <- table.freq(samples_with_mutation)
    composite_samples <- tbl$value[tbl$N > 1]
    singleton_samples <- tbl$value[tbl$N == 1]
    n_composite <- length(unique(strtrim(composite_samples, 9)))
    n_singleton <- length(unique(strtrim(singleton_samples, 9)))
    n_overall <- length(unique(strtrim(tbl$value, 9)))
    list(mutation=query.mutation,n_composite=n_composite, n_singleton=n_singleton, n_overall=n_overall)
}
l <- lapply(mutations, getN, tert)
l <- rbindlist(l)

info <- l[order(n_overall,decreasing=T),]
info <- info[n_overall >= 5,]
yinfo <- break_axis(info$n_overall,maxlower=60,minupper=500,lowerticksize=20,upperticksize=500,ratio_lower_to_upper=0.5)
info$y_overall <- yinfo$newy
info$y_composite <- info$y_overall * info$n_composite/info$n_overall
info$y_singleton <- info$y_overall * info$n_singleton/info$n_overall
infoM <- melt(info[,c('mutation','y_composite','y_singleton'),with=F], id.vars=c('mutation'))
infoM[variable=='y_composite',variable:='Composite']
infoM[variable=='y_singleton',variable:='Singleton']
infoM$variable <- factor(infoM$variable, levels=c('Singleton','Composite'))
infoM$mutation <- factor(infoM$mutation, levels=rev(info$mutation))
info$label <- paste0(round(info$n_composite / info$n_overall,3)*100,'%')
labels <- info[,c('mutation','label','y_composite'),with=F]
toremove <- which(yinfo$labels %in% c(500,1500,2500))
yinfo$newy <- yinfo$newy[-toremove]
yinfo$breaks <- yinfo$breaks[-toremove]
yinfo$labels <- yinfo$labels[-toremove]
yinfo$limits <- yinfo$limits[-toremove]

cols <- c('#EF7A63','#70C1DF')
names(cols) <- c('Composite','Singleton')
p1 <- ggplot(infoM, aes(x=mutation, y=value)) +
    geom_bar(stat='identity',fill='black') +
    scale_y_continuous(breaks=yinfo$breaks,labels=yinfo$labels,limits=yinfo$limits,expand=c(0,0),position='right') +
    theme_std(base_size=14) +
    labs(x=NULL,y='Number of patients') + 
    coord_flip()
p1 <- extract_gglegend(p1)

## format data for right plot showing the proportion of composite mutants including each mutation
## to expand the view in the lower proportions, we will plot from 0-25%, and then break axis to 100%
info2 <- info
info2$Composite <- info2$n_composite / info2$n_overall
info2$Singleton <- 1 - info2$Composite
info2 <- info2[,c('mutation','Composite','Singleton'),with=F]
info2 <- melt(info2, id.vars='mutation')
info2$mutation <- factor(info2$mutation, levels=rev(info$mutation))
info2$variable <- factor(info2$variable, levels=c('Singleton','Composite'))
info2[value >= 0.25,value:=value-0.75]

p2 <- ggplot(info2, aes(x=mutation, y=value)) +
    geom_bar(stat='identity',aes(fill=variable)) + 
    scale_y_continuous(expand=c(0,0),limits=c(0,0.27),breaks=c(0,0.1,0.25),labels=c(0,0.1,1),position='left') + 
    scale_fill_manual(values=cols,name=NULL) + 
    theme_std(base_size=14) +
    theme(axis.text.y=element_blank(),axis.ticks.y=element_blank()) + 
    labs(x=NULL,y='Fraction',title='') + 
    coord_flip()
p2 <- extract_gglegend(p2)
p <- plot_grid(p1$plot,p2$plot,nrow=1,ncol=2,align='h',rel_widths=c(4,1))
p <- plot_grid(p,p2$legend,nrow=2,ncol=1,rel_heights=c(4,1))
p

```

***

\pagebreak

### EDF-8b(ii)

```{r, panel_edf8bii, fig.width=6,fig.height=4}

all_samples <- d_clinical$Tumor_Sample_Barcode[d_clinical$exclude==F]
alleles <- info$mutation
x <- adt(expand.grid(m1=alleles, m2=alleles, stringsAsFactors=F))

test_comut <- function(i, x, d) {
    out <- x[i,]
    m1_samples <- d$Tumor_Sample_Barcode[d$mutation==out$m1]
    m2_samples <- d$Tumor_Sample_Barcode[d$mutation==out$m2]
    samples_AB <- all_samples[all_samples %nin% c(m1_samples, m2_samples)]
    samples_aB <- m1_samples[m1_samples %nin% m2_samples]
    samples_Ab <- m2_samples[m2_samples %nin% m1_samples]
    samples_ab <- intersect(m1_samples, m2_samples)
    AB <- length(unique(strsplit(samples_AB,9)))
    aB <- length(unique(strsplit(samples_aB,9)))
    Ab <- length(unique(strsplit(samples_Ab,9)))
    ab <- length(unique(strsplit(samples_ab,9)))
    m <- rbind(c(AB,aB),c(Ab,ab))
    tst <- fisher.test(m, alternative='two.sided')
    out$AB <- AB; out$aB <- aB; out$Ab <- Ab; out$ab <- ab
    out$OR <- as.numeric(tst$estimate)
    out$p <- as.numeric(tst$p.value)
    out
}


l <- lapply(1:nrow(x), test_comut, x, tert)
l <- rbindlist(l)
l$m1 <- factor(l$m1, levels=alleles)
l$m2 <- factor(l$m2, levels=alleles)
l$label <- l$ab
l$direction <- 'not significant'
l[p < 0.01 & OR < 1,direction:='mutually exclusive']
l[p < 0.01 & OR > 1,direction:='co-occurring']
l[p >= 0.01 & ab == 0, label:=NA]
l$logOR <- log2(l$OR)

m_ab <- reshape(l[,c('m1','m2','ab'),with=F],idvar='m1',timevar='m2',direction='wide')
rows <- m_ab$m1
m_ab[,m1:=NULL]
m_ab <- as.matrix(m_ab)
rownames(m_ab) <- rows
colnames(m_ab) <- rows
m_ab <- m_ab[alleles, alleles]
m_ab[upper.tri(m_ab,diag=T)] <- NA
mm_ab <- melt(m_ab)
names(mm_ab) <- c('m1','m2','ab')

m_label <- reshape(l[,c('m1','m2','label'),with=F],idvar='m1',timevar='m2',direction='wide')
rows <- m_label$m1
m_label[,m1:=NULL]
m_label <- as.matrix(m_label)
rownames(m_label) <- rows
colnames(m_label) <- rows
m_label <- m_label[alleles, alleles]
m_label[upper.tri(m_label,diag=T)] <- NA
mm_label <- melt(m_label)
names(mm_label) <- c('m1','m2','label')

m_direction <- reshape(l[,c('m1','m2','direction'),with=F],idvar='m1',timevar='m2',direction='wide')
rows <- m_direction$m1
m_direction[,m1:=NULL]
m_direction <- as.matrix(m_direction)
rownames(m_direction) <- rows
colnames(m_direction) <- rows
m_direction <- m_direction[alleles, alleles]
m_direction[upper.tri(m_direction,diag=T)] <- NA
mm_direction <- melt(m_direction)
names(mm_direction) <- c('m1','m2','direction')

mm <- adt(mm_ab)
mm$label <- mm_label$label
mm$direction <- mm_direction$direction
mm <- mm[m1 %nin% '228G>A' & m2 %nin% '205G>A']
mm$m1 <- factor(mm$m1, levels=alleles)
mm$m2 <- factor(mm$m2, levels=alleles)
mm[is.na(direction),direction:='not significant']

box_colors <- c('#EF7A63','black','white')
names(box_colors) <- c('mutually exclusive','co-occurring','not significant')
p <- ggplot(mm, aes(x=m2, y=m1)) +
    geom_tile(aes(fill=ab,color=direction),size=1,width=0.95,height=0.95) +
    scale_fill_gradient(low='#F1F1F2',high='steelblue',na.value='white',name='N') +
    scale_color_manual(values=box_colors,name=NULL) + 
    scale_x_discrete(expand=c(0,0)) +
    scale_y_discrete(expand=c(0,0)) +
    geom_text(aes(label=label)) + 
    theme_std(base_size=14) +
    theme(
        axis.text.x=element_text(angle=90,hjust=1,vjust=0.5),
        axis.ticks.x=element_blank(),
        axis.line.x=element_blank(),
        axis.ticks.y=element_blank(),
        axis.line.y=element_blank()) + 
    labs(x=NULL, y=NULL) 
p

```


