---
output:
    html_document:
        css: style.css
chunk_output_type: console
header-includes:
- \pagenumbering{gobble}
params:
    set_title: "Values in main text"
title: "`r params$set_title`"
---


```{r setup, include=F, echo=T}
knitr::opts_chunk$set(echo = F, message = F, warning = FALSE, autodep = knitr::dep_prev())

### source prerequisites (libraries, functions, variables)
source(here::here('r/prerequisites.R'))
d_clinical <- fread(here('data/data_clinical.txt.gz'))


### load starting datasets
d_mutations <- fread(here('data/data_mutations.txt.gz'),select=c('Tumor_Sample_Barcode','exclude','Hugo_Symbol','Role','putative_resistance_mutation',
                                                              'Variant_Classification','Variant_Type','tcn', 'mutationid', 'high_tmb','hotspot_class',
                                                              'tm','hotspotid','HGVSp_Short','truncating','Reference_Allele','Tumor_Seq_Allele2',
                                                              'Start_Position','End_Position'))
d_clinical_tcga <- fread(here('data/data_clinical_tcga.txt.gz'))
d_mutations_tcga <- fread(here('data/data_mutations_tcga.txt.gz'),select=c('Tumor_Sample_Barcode','Hugo_Symbol'))
gene_info <- fread(here('data/gene_info.txt'))
d_phased <- fread(here('data/data_mutations_phased.txt.gz')) 
```


## Paragraph 1

### Number of patients in study:
```{r, echo=T}
length(unique(strtrim(d_clinical$Tumor_Sample_Barcode,9)))
```

# Selection for composite mutations

## Paragraph 2


### Percentage of tumors harboring a composite mutation in an IMPACT gene:
```{r, echo=T}
n_samples <- length(unique(d_clinical$Tumor_Sample_Barcode))
check_composites_per_sample <- function(d) {
    any_composite <- any(duplicated(d$Hugo_Symbol))
    list(any_composite=any_composite)
}
result <- d_mutations[,check_composites_per_sample(.SD),by=Tumor_Sample_Barcode]
samples_with_composites <- result$Tumor_Sample_Barcode[result$any_composite==T]
n_samples_with_composites <- length(unique(samples_with_composites))
fraction_samples_with_composites <- n_samples_with_composites / n_samples
round(100*fraction_samples_with_composites,3)
```

### Number of tumor specimens harboring any composite mutations
```{r, echo=T}
n_samples_with_composites
```

### Expected percentage of tumor specimens harboring any composite mutations by chance
```{r, echo=T}
### load the percent of composites across all samples after each of 100,000 random permutations
### the average is the expected percentage of composite mutations by chance
l <- readRDS(here('data/observed_vs_expected_composites_impact_precalculated.rds'))
average_expected <- mean(l$dat_all$prop)
average_expected
```

### Percent increase of observed to expected rate of composite-mutated samples in data
```{r, echo=T}
round(100*(fraction_samples_with_composites / average_expected)-100,1)

```

### Mutation burdens with enriched composite mutations
```{r,echo=T}
l <- readRDS(here('data/observed_vs_expected_composites_per_tmb_impact_precalculated.rds'))
get_p <- function(x) {
    pvalue <- x$p
    tmb <- x$burden
    list(tmb=tmb,pvalue=pvalue)
}
result <- rbindlist(lapply(l, get_p))
most_enriched <- result$tmb[result$pvalue < 1e-5]
data.table(TMB=most_enriched,pvalue='<1e-5')
```

### Percentage of all samples with TMB: 4-11
```{r, echo=T}
x <- sum(round(d_clinical$tmb) %in% 4:11)
round(100*(x/n_samples),3)
```

#### Percentage of composite-mutant cases that could be ascribed to resistance mutations
```{r, echo=T}
summarize_sample_gene <- function(d) {
    composite <- nrow(d) > 1
    any_resistance <- any(d$putative_resistance_mutation,na.rm=T) & composite==T
    excluded <- any(d$exclude==T)
    if(composite==F) {
        origin <- 'n/a'
    } else if(excluded==T) {
        origin <- 'Hypermutation/High-TMB'
    } else if(any_resistance) {
        origin <- 'Rx resistance'
    } else {
        origin <- 'Unclear'
    }
    list(composite=composite,origin=origin)
}
info <- d_mutations[,summarize_sample_gene(.SD),by=c('Tumor_Sample_Barcode','Hugo_Symbol')]

summarize_sample <- function(d) {
    any.composite <- any(d$composite)
    excluded <- any(d$origin=='Hypermutation/High-TMB')
    any.resistance <- any(d$origin=='Rx resistance')
    if(any.composite==F) {
        reason <- 'n/a'
    } else if(excluded) {
        reason <- 'Hypermutation/High-TMB'
    } else if(any.resistance) {
        reason <- 'Rx resistance'
    } else {
        reason <- 'Unclear'
    }
    list(any.composite=any.composite, reason=reason)
}
sample_result <- info[,summarize_sample(.SD),by=Tumor_Sample_Barcode]
sample_result <- sample_result[any.composite==T,]
round(sum(sample_result$reason=='Rx resistance')/nrow(sample_result),3)
```



## Paragraph 3

### Percentage of cases affected with composite mutations in oncogenes and TSGs
```{r, echo=T}

## load mutation data
d <- d_mutations[,c('exclude','putative_resistance_mutation','Tumor_Sample_Barcode','Hugo_Symbol','Role'),with=F]
d <- d[exclude==F & putative_resistance_mutation==F,]
d$id <- paste(d$Tumor_Sample_Barcode,d$Hugo_Symbol)
tbl <- table.freq(d$id)
tbl$composite <- tbl$N > 1
tbl[,N:=NULL]
d <- merge(d, tbl, by.x='id', by.y='value', all.x=T)
clin <- fread(here('data/data_clinical.txt.gz'))
clin <- clin[exclude==F,]
N_patients <- length(unique(clin$PATIENT_ID))
tabulate_composites <- function(d) {
    comp_onc <- any(d$composite==T & d$Role=='Oncogene')
    comp_tsg <- any(d$composite==T & d$Role=='TSG')
    list(comp_onc=comp_onc, comp_tsg=comp_tsg)
}
info <- d[,tabulate_composites(.SD),by=Tumor_Sample_Barcode]
n_comp_onc <- length(unique(strtrim(info$Tumor_Sample_Barcode[info$comp_onc==T],9)))
n_comp_tsg <- length(unique(strtrim(info$Tumor_Sample_Barcode[info$comp_tsg==T],9)))
n_sing_onc <- N_patients - n_comp_onc
n_sing_tsg <- N_patients - n_comp_tsg
ci_onc <- binom.confint(n_comp_onc, N_patients, methods='exact')
ci_tsg <- binom.confint(n_comp_tsg, N_patients, methods='exact')
ci_onc$Role <- 'Oncogenes'
ci_tsg$Role <- 'TSGs'
ci <- adt(rbind(ci_onc, ci_tsg))
ci[,c('mean','lower','upper'):=list(100*mean,100*lower,100*upper)]
ci

## p-value based on mc.nemar's test (paired samples with 2 binary features)
m <- rbind(c(n_sing_onc,n_comp_onc),c(n_sing_tsg,n_comp_tsg))
p <- mcnemar.test(m)$p.value
pval_lab <- paste0('P~',prettyNum(p,digits=1))
pval_lab

```

### Percentage of composite mutations in TSGs or Oncogenes with 1+ truncating variants vs only missense
```{r, echo=T}

### load phased data with pairwise-mutation comparison
d <- d_phased

### annotate class (i.e. missense/missense, missense/truncating, etc.
d <- d[exclude==F & putative_resistance_mutation.1==F & putative_resistance_mutation.2==F,]
d$Class <- ''
d[truncating.1==F & truncating.2==F,Class:='Non-truncating/Non-truncating']
d[(truncating.1==T & truncating.2==F | truncating.1==F & truncating.2==T),Class:='Non-truncating/Truncating']
d[truncating.1==T & truncating.2==T,Class:='Truncating/Truncating']
d <- d[Role %in% c('Oncogene','TSG'),]

### don't overcount 3+ mutation composite-mutations by using unique sample/gene/Class counts
d$id <- paste(d$Tumor_Sample_Barcode, d$Hugo_Symbol, d$Class)
d <- d[!duplicated(id),c('Tumor_Sample_Barcode','Hugo_Symbol','Class','Role','id'),with=F]
dd <- d
dd[Class!='Non-truncating/Non-truncating',Class:='1 or both are truncating']
dd$Class <- factor(dd$Class,levels=(c('Non-truncating/Non-truncating','1 or both are truncating')))

### for each gene, get its singleton/composite counts
summarize_gene <- function(d) {
    n_missense_missense <- length(unique(strtrim(d$Tumor_Sample_Barcode[d$Class=='Non-truncating/Non-truncating'],9)))
    n_any_truncating <- length(unique(strtrim(d$Tumor_Sample_Barcode[d$Class=='1 or both are truncating'],9)))
    tbl <- list(`Non-truncating/Non-truncating`=n_missense_missense, `1 or both are truncating`=n_any_truncating)
    tbl
}
res <- dd[,summarize_gene(.SD),by=c('Role')]
m <- as.matrix(res[,2:3,with=F])
rownames(m) <- res$Role
round(100*(m / rowSums(m)),3)
```


## Paragraph 4

### Genes significantly enriched for composite mutations
```{r, echo=T}
d <- fread(here('data/gene_enrichment_precalculated.txt'))
d <- d[q_enriched < 0.01,]
length(d$Hugo_Symbol)
```

### Percentage of PIK3CA variants that were composite mutations
```{r, echo=T}
composites <- d$composites[d$Hugo_Symbol=='PIK3CA']
affected <- d$samples[d$Hugo_Symbol=='PIK3CA']
ci <- binom::binom.confint(composites, affected, method='exact')
round(100*ci[,4:6],1)
```


## Paragraph 5

### How many times more likely are composite mutations to include hotspots than singleton mutations?
```{r, echo=T}

### load MAF, subset for hotspot SNPs among valid samples
d <- d_mutations
d <- d[exclude==F & putative_resistance_mutation==F,]
dd <- d[Variant_Classification %in% 'Missense_Mutation',]
dd[hotspot_class=='3d',hotspot:=F]
dd[hotspot_class=='3d',hotspot_class:='']

### annotate if the sample-gene is singleton or composite
dd$id <- paste(dd$Tumor_Sample_Barcode,dd$Hugo_Symbol)
tbl <- table.freq(dd$id)
singleton <- tbl$value[tbl$N==1]
composite <- tbl$value[tbl$N>1]
dd$composite <- dd$id %in% composite
dd$PATIENT_ID <- strtrim(dd$Tumor_Sample_Barcode,9)
dd$pid_gene <- paste(dd$PATIENT_ID,dd$Hugo_Symbol)
total_number_singleton_mutations <- length(unique(dd$pid_gene[dd$composite==F]))
total_number_composite_mutations <- length(unique(dd$pid_gene[dd$composite==T]))
total_number_mutations <- length(unique(dd$pid_gene))

### for each hotspot, count the number of mutant, singleton and composite samples
summarize_individual_hotspot <- function(d) {
    ### count affected singleton/composite samples
    mutant_samples <- sortunique(d$Tumor_Sample_Barcode)
    singleton_samples <- sortunique(d$Tumor_Sample_Barcode[d$composite==F])
    composite_samples <- sortunique(d$Tumor_Sample_Barcode[d$composite==T])

    ### don't overcount patients with 2+ samples
    mutant_samples <- unique(strtrim(mutant_samples,9))
    singleton_samples <- unique(strtrim(singleton_samples,9))
    composite_samples <- unique(strtrim(composite_samples,9))

    n.mutant <- length(mutant_samples)
    n.singletons <- length(singleton_samples)
    n.composites <- length(composite_samples)
    list(n.mutant=n.mutant,n.singletons=n.singletons,n.composites=n.composites)
}

hotspots <- sortunique(dd$hotspotid)
hotspots <- hotspots[hotspots!='']
ll <- dd[hotspotid!='',summarize_individual_hotspot(.SD),by=hotspotid]
ll <- ll[order(n.mutant,decreasing=T),]
ll$x.order <- 1:nrow(ll)
ll$p.singletons <- cumsum(ll$n.singletons / total_number_singleton_mutations)
ll$p.composites <- cumsum(ll$n.composites / total_number_composite_mutations)
ll$p.mutant <- cumsum(ll$n.mutant / total_number_mutations)

overall_prob_composite_mutation_has_hotspot <- max(ll$p.composites)
overall_prob_singleton_mutation_has_hotspot <- max(ll$p.singleton)
overall_prob_composite_mutation_has_hotspot / overall_prob_singleton_mutation_has_hotspot

### Calculate P-value:
singletons_summary <- data.table(x=sum(ll$n.singletons),n=total_number_singleton_mutations,class='Singletons')
composites_summary <- data.table(x=sum(ll$n.composites),n=total_number_composite_mutations,class='Composite mutations')
summary <- rbind(singletons_summary, composites_summary)
prop.test(summary$x, summary$n)$p.value
```


### Percentage of composite mutations per gene role where more-prevalent allele is mutated first
```{r, echo=T}

### load phasing data
d <- d_phased
d <- d[putative_resistance_mutation.1==F & putative_resistance_mutation.2==F & exclude==F,]

## annotate functional mutations
d[hotspotid.1!='',tm.1:=hotspotid.1] ### this will group IF-INDEL hotspots
d[hotspotid.2!='',tm.2:=hotspotid.2]
d$functional.1 <- d$hotspotid.1!='' | d$oncogenic.1 %in% c('Oncogenic','Likely Oncogenic','Predicted Oncogenic') |
(d$Role %in% c('TSG','Oncogene/TSG') & d$truncating.1==T) | d$Variant_Classification.1=='TERT promoter'
d$functional.2 <- d$hotspotid.2!='' | d$oncogenic.2 %in% c('Oncogenic','Likely Oncogenic','Predicted Oncogenic') |
(d$Role %in% c('TSG','Oncogene/TSG') & d$truncating.2==T) | d$Variant_Classification.2=='TERT promoter'
d[Variant_Classification.1 %in% c('TERT promoter'),tm.1:=paste0('TERT ',gsub('p[.]','',HGVSp_Short.1))]
d[Variant_Classification.1 %in% c('Splice_Site','Nonsense_Mutation','Frame_Shift_Del','Frame_Shift_Ins'), tm.1:=paste(Hugo_Symbol,'Truncating')]
d[Variant_Classification.2 %in% c('TERT promoter'),tm.2:=paste0('TERT ',gsub('p[.]','',HGVSp_Short.2))]
d[Variant_Classification.2 %in% c('Splice_Site','Nonsense_Mutation','Frame_Shift_Del','Frame_Shift_Ins'), tm.2:=paste(Hugo_Symbol,'Truncating')]
d <- d[,c('Tumor_Sample_Barcode','Role','Hugo_Symbol','hotspot.1','hotspot.2','functional.1','functional.2',
        'tm.1','tm.2','phase','common_reads_alt1_alt2',
        'common_reads_alt1_ref2','common_reads_ref1_alt2','ccf_Mcopies_lower.1','ccf_Mcopies_upper.1',
        'ccf_Mcopies_lower.2','ccf_Mcopies_upper.2','Variant_Classification.1','Variant_Classification.2','phase','same_cell_known'),with=F]

### annotate which allele was mutated first where this can be determined
d$firstallele <- ''
d$secondallele <- ''
d[ccf_Mcopies_lower.1 > ccf_Mcopies_upper.2,firstallele:=tm.1]
d[ccf_Mcopies_lower.1 > ccf_Mcopies_upper.2,secondallele:=tm.2]
d[ccf_Mcopies_upper.1 < ccf_Mcopies_lower.2,firstallele:=tm.2]
d[ccf_Mcopies_upper.1 < ccf_Mcopies_lower.2,secondallele:=tm.1]
dd <- d[,c('Tumor_Sample_Barcode','Role','Hugo_Symbol','firstallele','secondallele','phase',
           'functional.1','functional.2',
           'same_cell_known','hotspot.1','hotspot.2',
           'Variant_Classification.1','Variant_Classification.2'),with=F]
dd <- dd[firstallele!='' & secondallele!='',]

### get number of unique patients with each mutation
d0 <- d_mutations
d0 <- d0[exclude==F & putative_resistance_mutation==F,]
d0$functional <- d0$hotspot==T | d0$oncogenic %in% c('Oncogenic','Likely Oncogenic','Predicted Oncogenic') |
(d0$Role %in% c('TSG','Oncogene/TSG') & d0$truncating==T) | d0$Variant_Classification=='TERT promoter'
d0[Variant_Classification %in% c('TERT promoter'),tm:=paste0('TERT ',gsub('p[.]','',HGVSp_Short))]
d0[Variant_Classification %in% c('Splice_Site','Nonsense_Mutation','Frame_Shift_Del','Frame_Shift_Ins'), tm:=paste(Hugo_Symbol,'Truncating')]
d0[hotspotid!='',tm:=hotspotid] ### this will group IF-INDEL hotspots
getN <- function(d0) list(N=length(unique(strtrim(d0$Tumor_Sample_Barcode,9))))
tbl <- d0[,getN(.SD),by=tm]
setnames(tbl,'tm','value')

### annotate first and second hits with their prevalence across full dataset
dd2 <- merge(dd, tbl, by.x='firstallele', by.y='value', all.x=T)
setnames(dd2,'N','firstallele_N')
dd2 <- merge(dd2, tbl, by.x='secondallele', by.y='value', all.x=T)
setnames(dd2,'N','secondallele_N')
tbl_gene <- table.freq(d0$Hugo_Symbol)
dd2 <- merge(dd2, tbl_gene, by.x='Hugo_Symbol', by.y='value', all.x=T)
setnames(dd2,'N','totalN')

### compare prevalence of each hit for signficantly more prevalent:
first_ci <- binom::binom.confint(dd2$firstallele_N, dd2$totalN, method='exact')
second_ci <- binom::binom.confint(dd2$secondallele_N, dd2$totalN, method='exact')
dd2$prop_samples_with_first_allele <- first_ci$mean
dd2$prop_samples_with_first_allele_lwr <- first_ci$lower
dd2$prop_samples_with_first_allele_upr <- first_ci$upper
dd2$prop_samples_with_second_allele <- second_ci$mean
dd2$prop_samples_with_second_allele_lwr <- second_ci$lower
dd2$prop_samples_with_second_allele_upr <- second_ci$upper
dd2$more_common_allele <- 'no difference'
dd2[prop_samples_with_first_allele_lwr > prop_samples_with_second_allele_upr, more_common_allele:='first']
dd2[prop_samples_with_first_allele_upr < prop_samples_with_second_allele_lwr, more_common_allele:='second']
dd2$Hugo_Symbol <- as.character(dd2$Hugo_Symbol)
dd2$firstallele <- as.character(dd2$firstallele)
dd2$secondallele <- as.character(dd2$secondallele)
dat <- dd2

### get the prop of 1st more prevalent vs 2nd more prevalent vs gene role
dat <- dat[same_cell_known==T | phase=='cis',] ### only consider composites definitely in same cell
dat$n_functional <- dat$functional.1 + dat$functional.2
dat <- dat[n_functional > 0,]
dat$n_functional <- paste0('functional_',dat$n_functional)
dat$n_functional <- factor(dat$n_functional)
dat <- dat[more_common_allele!='no difference',]
m <- as.data.frame(t(as.data.frame.matrix(xtabs(~more_common_allele + n_functional, data=dat[Role=='Oncogene']))))
onc_propdata <- adt(binom::binom.confint(m$first,(m$first+m$second),method='exact'))
onc_propdata$n_functional <- c(1,2)
onc_propdata$role <- 'Oncogenes'
m <- as.data.frame(t(as.data.frame.matrix(xtabs(~more_common_allele + n_functional, data=dat[Role=='TSG']))))
tsg_propdata <- adt(binom::binom.confint(m$first,(m$first+m$second),method='exact'))
tsg_propdata$n_functional <- c(1,2)
tsg_propdata$role <- 'TSGs'
propdata <- rbind(onc_propdata, tsg_propdata)
propdata <- propdata[n_functional==2,]
setnames(propdata,'mean','prop_first')
out <- propdata[,c('role','prop_first','lower','upper'),with=F]
out[,2:4] <- 100*out[,(2:4),with=F]
names(out) <- c('Role','% more prevalent mutated first','95%CI (lower','95%CI (upper)')

### Percentage of composite mutations per gene role where more-prevalent allele is mutated first
out
```



# Phase and function

## Paragraph 6

### Tabulate phase-able composite mutations
```{r, echo=T}
d <- d_phased[exclude==F & !is.na(phase) & phase!='unknown',]

### count composite mutations with 3+ mutations of the same phase only once
d$id <- paste(d$Tumor_Sample_Barcode,d$Hugo_Symbol,d$phase)
d <- d[!duplicated(id),]

### tabulate composites with different phasing outcomes
d[phase %in% c('trans or separate cells','ambiguous: 3+ cis, 3+ trans'),phase:='ambiguous']
tbl <- table.freq(d$phase)
tbl$Percentage <- round(100*(tbl$N / sum(tbl$N)),3)
setnames(tbl,'value','Phase')
tbl
```

### Compare cis/trans rates in TSGs and Oncogenes with a) 2 missense-mutations, b) 1+ truncating mutations
```{r, echo=T}

## load phased data (using the pairwise-combos)
d <- d_phased

d <- d[exclude==F & putative_resistance_mutation.1==F & putative_resistance_mutation.2==F,]
d$Class <- ''
d[truncating.1==F & truncating.2==F,Class:='Non-truncating/Non-truncating']
d[(truncating.1==T & truncating.2==F | truncating.1==F & truncating.2==T),Class:='Non-truncating/Truncating']
d[truncating.1==T & truncating.2==T,Class:='Truncating/Truncating']

### don't overcount 3+ mutation composite-mutations; using unique sample/gene/Class counts
d$id <- paste(d$Tumor_Sample_Barcode, d$Hugo_Symbol, d$Class, d$phase)
dd <- d
dd <- dd[!duplicated(id),c('Tumor_Sample_Barcode','Hugo_Symbol','Class','Role','phase','id'),with=F]

### format the Classes: Oncogene/nontrunc/nontrunc vs TSG (3 categories)
dd$Class <- paste0(dd$Role,' (',dd$Class,')')
dd <- dd[Role=='TSG' | Class %in% c('Oncogene (Non-truncating/Non-truncating)'),]
class_levels <- sortunique(dd$Class)
dd$Class <- factor(dd$Class,levels=class_levels)
dd$phase <- factor(dd$phase,levels=c('cis','trans'))

### tabulate the results
dat <- as.data.frame.matrix(xtabs(~Class + phase,data=dd))
dat$N <- dat$cis + dat$trans
dat <- cbind(Class=rownames(dat),adt(dat))
dat$Class <- factor(dat$Class,levels=dat$Class)
dat$prop_cis <- dat$cis / (dat$cis + dat$trans)
dat$prop_trans <- 1 - dat$prop_cis

### P-value
m <- dat[Class %in% c('Oncogene (Non-truncating/Non-truncating)','TSG (Truncating/Truncating)'),]
classes <- m$Class
m <- as.matrix(m[,(2:3),with=F])
rownames(m) <- classes
p=fisher.test(m,alternative='two.sided')$p.value

## breakdown of cis/trans vs missense+missense or 1+truncating
dat
p
```

### P-value for significance of 100% resistance mutation composites in cis vs non-resistance
```{r, echo=T}

## here we will find composite mutations associated with resistance, tabulate their 
## incidence in cis and trans, and compare this to the cis/trans frequencies for 
## composites not including resistance mutations.
## NB: Phased inferred from LOH was used in the figure panel
## "Fraction of phase-able composite mutations" per gene, but not to calculate 
## P-value for difference with non-resistance composites, 
## as this inferrential phasing was not generally done for them.

## define helper functions here
label_composite <- function(d) {
    label <- ''
    if(d$Hugo_Symbol=='EGFR' & (d$HGVSp_Short.1=='Exon 19 INDELs' & d$HGVSp_Short.2=='T790M'))  label <- 'EGFR Exon 19 INDELs + T790M'
    if(d$Hugo_Symbol=='EGFR' & (d$HGVSp_Short.1=='T790M' & d$HGVSp_Short.2=='L858R'))  label <- 'EGFR L858R + T790M'
    if(d$Hugo_Symbol=='EGFR' & (d$HGVSp_Short.1=='D761Y' & d$HGVSp_Short.2=='T790M'))  label <- 'EGFR D761Y + T790M'
    if(d$Hugo_Symbol=='EGFR' & (d$HGVSp_Short.1=='G719A' & d$HGVSp_Short.2=='T790M'))  label <- 'EGFR G719A + T790M'
    if(d$Hugo_Symbol=='EGFR' & (d$HGVSp_Short.1=='T790M' & d$HGVSp_Short.2=='C797S'))  label <- 'EGFR T790M + C797S'
    if(d$Hugo_Symbol=='EGFR' & (d$HGVSp_Short.1=='Exon 19 INDELs' & d$HGVSp_Short.2=='C797S'))  label <- 'EGFR Exon 19 INDELs + C797S'
    if(d$Hugo_Symbol=='EGFR' & (d$HGVSp_Short.1=='C797S' & d$HGVSp_Short.2=='L858R'))  label <- 'EGFR L858R + C797S'
    if(d$Hugo_Symbol=='EGFR' & (d$HGVSp_Short.1=='D761Y' & d$HGVSp_Short.2=='C797S'))  label <- 'EGFR D761Y + C797S'
    if(d$Hugo_Symbol=='EGFR' & (d$HGVSp_Short.1=='G719A' & d$HGVSp_Short.2=='C797S'))  label <- 'EGFR G719A + C797S'
    if(d$Hugo_Symbol=='AR' &   (d$HGVSp_Short.1=='L702H'  & d$HGVSp_Short.2=='H875Y'))  label <- 'AR L702H + H875Y'
    if(d$Hugo_Symbol=='ALK' &  (d$HGVSp_Short.1=='D1203N' & d$HGVSp_Short.2=='L1196M')) label <- 'ALK D1203N + L1196M'
    if(d$Hugo_Symbol=='ALK' &  (d$HGVSp_Short.1=='G1202R' & d$HGVSp_Short.2=='L1196M')) label <- 'ALK G1202R + L1196M'
    if(d$Hugo_Symbol=='BRCA1' &  (d$HGVSp_Short.1=='E1257Gfs*9' & d$HGVSp_Short.2=='N1235Tfs*18')) label <- 'BRCA1 Frame-restoring INDELs'
    if(label=='')  label <- paste(d$Hugo_Symbol,d$HGVSp_Short.1,'+',d$HGVSp_Short.2)
    d$label <- label
    d
}

extract_cis_trans <- function(d) {
    ### don't overcount multiple pairs of composite mutations in the same sample+gene+phase
    d$id <- paste(d$Tumor_Sample_Barcode,d$Hugo_Symbol,d$phase)
    d <- d[!duplicated(id),c('Tumor_Sample_Barcode','metamaintype','Hugo_Symbol','Role','phase','id',
                             'putative_resistance_mutation.1','putative_resistance_mutation.2',
                             'HGVSp_Short.1','HGVSp_Short.2',
                             'ccf_Mcopies.1','clonal.1','ccf_Mcopies.2','clonal.2',
                             'common_reads_alt1_alt2','common_reads_alt1_ref2','common_reads_ref1_alt2','common_reads_ref1_ref2',
                             't_depth.1','t_depth.2'),with=F]
    d
}

summarize_composite_with_alleles <- function(d) {
    n <- length(unique(strtrim(d$Tumor_Sample_Barcode,9)))
    list(n=n)
}


### load phased data, add phasing by LOH with FACETS and manual review
d <- d_phased
d <- d[exclude==F & (putative_resistance_mutation.1==T | putative_resistance_mutation.2==T),]
reviewed <- fread(here('data/resistance_mutation_phasing_reviewed.txt'))
reviewed <- reviewed[,c('Tumor_Sample_Barcode','Hugo_Symbol','HGVSp_Short.1','HGVSp_Short.2','phase_reviewed'),with=F]
reviewed$id <- paste(reviewed$Tumor_Sample_Barcode,reviewed$Hugo_Symbol,reviewed$HGVSp_Short.1,reviewed$HGVSp_Short.2)
reviewed <- reviewed[!duplicated(id),]
d <- merge(d, reviewed[,c('Tumor_Sample_Barcode','Hugo_Symbol','HGVSp_Short.1','HGVSp_Short.2','phase_reviewed'),with=F],
           by=c('Tumor_Sample_Barcode','Hugo_Symbol','HGVSp_Short.1','HGVSp_Short.2'),all.x=T)
d[phase_reviewed %in% c('cis','cis*'),phase:=phase_reviewed]

### exclude these composite mutations for stated reasons
to_remove <- c(
               ### bad facets fit in this sample, we can't be sure variants are in the same cell: exclude
               which(d$Tumor_Sample_Barcode=='P-0013110-T01-IM5' & d$tm.1=='KIT C809' & d$tm.2=='KIT indel:820-821'),
               ### this sample has EGFR S768 cis with EGFR T790, but both are in trans with a passenger C775, so I am removing the passenger composite
               which(d$Tumor_Sample_Barcode=='P-0003923-T01-IM5' & d$tm.1=='EGFR C775' & d$tm.2=='EGFR T790'),
               which(d$Tumor_Sample_Barcode=='P-0003923-T01-IM5' & d$tm.1=='EGFR S768' & d$tm.2=='EGFR C775'))
d <- d[-to_remove,]

### collapse in-frame INDEL mutation labels to just the exon of the INDEL
d$HGVSp_Short.1 <- gsub('p[.]','',d$HGVSp_Short.1)
d$HGVSp_Short.2 <- gsub('p[.]','',d$HGVSp_Short.2)
d[Variant_Classification.1 %in% c('In_Frame_Ins','In_Frame_Del') & Hugo_Symbol=='EGFR' & Exon_Number.1=='19/28',HGVSp_Short.1:=paste('Exon 19 INDELs')]
d[Variant_Classification.1 %in% c('In_Frame_Ins','In_Frame_Del') & Hugo_Symbol=='EGFR' & Exon_Number.1=='20/28',HGVSp_Short.1:=paste('Exon 20 INDELs')]
d[Variant_Classification.2 %in% c('In_Frame_Ins','In_Frame_Del') & Hugo_Symbol=='EGFR' & Exon_Number.2=='19/28',HGVSp_Short.2:=paste('Exon 19 INDELs')]
d[Variant_Classification.2 %in% c('In_Frame_Ins','In_Frame_Del') & Hugo_Symbol=='EGFR' & Exon_Number.2=='20/28',HGVSp_Short.2:=paste('Exon 20 INDELs')]
d[Variant_Classification.1 %in% c('In_Frame_Ins','In_Frame_Del') & Hugo_Symbol=='KIT' & Exon_Number.1=='11/21',HGVSp_Short.1:=paste('Exon 11 INDELs')]
d[Variant_Classification.2 %in% c('In_Frame_Ins','In_Frame_Del') & Hugo_Symbol=='KIT' & Exon_Number.2=='11/21',HGVSp_Short.2:=paste('Exon 11 INDELs')]
d <- d[phase %in% c('cis','cis*','trans'),]

### don't overcount composites within each sample/gene/alleles/phase combination (redoing after the above label collapsing in last step)
d$id <- paste(strtrim(d$Tumor_Sample_Barcode,9),d$Hugo_Symbol,d$HGVSp_Short.1,d$HGVSp_Short.2,d$phase)
d <- d[!duplicated(id),]

### format data
dd <- extract_cis_trans(d)
dd <- dd[order(Tumor_Sample_Barcode,Hugo_Symbol),]
dd$i <- 1:nrow(dd)
dd <- dd[,label_composite(.SD),by=i]

### these samples have bona fide EGFR resistance mutation with 2+ other EGFR mutations, but the resistance mutation was not phaseable with any, so exclude them
toremove <- which(dd$Hugo_Symbol=='EGFR' & (grepl('T790M',dd$label)==F & grepl('C797S',dd$label)==F))
dd <- dd[-toremove,]

### Exclude ESR1 resistance mutations because ESR1 there is no initial ESR1 drug target
dd <- dd[Hugo_Symbol!='ESR1',]

### get final tabulation of phaseable composite mutations associated with therapeutic resistance
info <- dd[,summarize_composite_with_alleles(.SD),by=c('Hugo_Symbol','label','phase')]
info <- info[phase %in% c('cis','cis*','trans'),]
summarize_gene <- function(info) list(n=sum(info$n))
ns <- info[,summarize_gene(.SD),by=Hugo_Symbol]
ns <- ns[order(n,decreasing=T),]
info$Hugo_Symbol <- factor(info$Hugo_Symbol, levels=ns$Hugo_Symbol)
info <- info[order(info$Hugo_Symbol,info$n,decreasing=F),]
info$label <- factor(info$label, levels=info$label)

### get the number of cis/trans composite mutations not associated with resistance
d <- d_phased
d <- d[exclude==F & (putative_resistance_mutation.1==F & putative_resistance_mutation.2==F),]
d$important.1 <- d$hotspot.1==T | d$oncogenic.1 %in% c('Oncogenic','Likely Oncogenic','Predicted Oncogenic') |
    (d$Role %in% c('TSG','Oncogene/TSG') & d$truncating.1==T)
d$important.2 <- d$hotspot.2==T | d$oncogenic.2 %in% c('Oncogenic','Likely Oncogenic','Predicted Oncogenic') |
    (d$Role %in% c('TSG','Oncogene/TSG') & d$truncating.2==T)
dd <- extract_cis_trans(d)
dd$id <- paste(strtrim(dd$Tumor_Sample_Barcode,9),dd$Hugo_Symbol,dd$HGVSp_Short.1,dd$HGVSp_Short.2,dd$phase)
dd <- dd[!duplicated(id),]
dd <- dd[order(Hugo_Symbol),]
dd$i <- 1:nrow(dd)
tbl <- table.freq(dd$phase)
tbl <- tbl[value %in% c('cis','trans'),]
names(tbl) <- c('phase','N')
tbl$group <- 'All non-resistance mutations'

### merge the result with the values for resistance mutations calculated above
tbl_rm <- data.table(phase=c('cis','trans'),N=c(sum(info$n[info$phase=='cis']),0),group='Resistance mutations')
dat <- rbind(tbl, tbl_rm)

# compare cis and trans rates between composites associated with/without resistance
m0 <- as.data.frame(reshape(dat,idvar='group',timevar='phase',direction='wide'))
rownames(m0) <- m0$group
m0$group <- NULL
names(m0) <- gsub('N[.]','',names(m0))
pval <- fisher.test(m0)$p.value
m0
pval
```


## Paragraph 7

### Percentage of cis composites among oncogenes with greatest cis rate
```{r, echo=T}
### count the instances of cis/trans per gene
d <- d_phased
d <- d[exclude==F & putative_resistance_mutation.1==F & putative_resistance_mutation.2==F,]
d$important.1 <- d$hotspot.1==T | d$oncogenic.1 %in% c('Oncogenic','Likely Oncogenic','Predicted Oncogenic') | (d$Role %in% c('TSG','Oncogene/TSG') & d$truncating.1==T)
d$important.2 <- d$hotspot.2==T | d$oncogenic.2 %in% c('Oncogenic','Likely Oncogenic','Predicted Oncogenic') | (d$Role %in% c('TSG','Oncogene/TSG') & d$truncating.2==T)
dd <- d[important.1==T | important.2==T,]
dd <- extract_cis_trans(dd)

### get the number of patients with cis or trans composites in each gene
quantify_gene <- function(d) {
    cis_unique_samples <- length(unique(strtrim(d$Tumor_Sample_Barcode[d$phase=='cis'],9)))
    trans_unique_samples <- length(unique(strtrim(d$Tumor_Sample_Barcode[d$phase=='trans'],9)))
    list(n.cis=cis_unique_samples,n.trans=trans_unique_samples)
}
result <- dd[,quantify_gene(.SD),by=c('Hugo_Symbol','Role')]
result$percent_trans <- result$n.trans / (result$n.cis + result$n.trans)
result$percent_cis <- 1 - result$percent_trans
result <- result[order(percent_cis,decreasing=T),]
result$total_phaseable <- result$n.cis + result$n.trans
result <- result[total_phaseable >= 10,]

## load results from gene-enrichment test, exclude genes with possible depletion
gene_enrichment <- fread(here('data/gene_enrichment_precalculated.txt'))
depleted_genes <- gene_enrichment$Hugo_Symbol[gene_enrichment$q_twosided < 0.01 & gene_enrichment$logOR < 0]
result <- result[!Hugo_Symbol %in% depleted_genes,]

### show results for oncogenes with decreasing cis proportion
result[Role=='Oncogene']

```

### prevalent cis-acting composite mutations were observed even among canonical TSGs
```{r, echo=T}
d <- d_phased
d <- d[exclude==F & putative_resistance_mutation.1==F & putative_resistance_mutation.2==F,]
tbl <- table.freq(d$phase[d$Role=='TSG' & d$phase %in% c('cis','trans')])
tbl$prop <- tbl$N/sum(tbl$N)
tbl
```

### TP53 was notable among TSGs
```{r, echo=T}
result[Hugo_Symbol=='TP53']
```


# Conditionally dependent mutant alleles

## Paragraph 9

### Residues enriched for composite mutations
```{r,echo=T}
    d <- fread(here('data/residue_enrichment_precalculated.txt'))
    length(unique(d$query_mutation[d$q.value.enriched < 0.01])) 
```

### Genes with any enriched for composite mutations
```{r,echo=T}
    length(unique(d$query_gene[d$q.value.enriched < 0.01])) 
```

### Percentage of enriched mutations among TP53, PIK3CA, APC, and EGFR
```{r,echo=T}
    genes <- c('TP53','PIK3CA','APC','EGFR')
    x <- d[q.value.enriched < 0.01]    
    round(100*sum(x$query_gene %in% genes) / nrow(x), 3)
```

### Percentage of enriched mutations that are also hotspots
```{r,echo=T}
    enriched_residues <- x$query_mutation
    sum(x$hotspotid!='')/nrow(x) ### hotspotid are from 24K analysis
```

### PIK3CA residues enriched for composites are almost always in cis, and always arise as missense mutations
```{r, echo=T}
tmp <- x[query_gene=='PIK3CA']
tmp <- tmp[order(q.value.enriched,decreasing=F),]
out <- tmp[,c('query_mutation','n_cis','n_trans','q.value.enriched'),with=F]

dd <- d_phased[exclude==F & Hugo_Symbol=='PIK3CA', c('Tumor_Sample_Barcode','tm.1','tm.2','Variant_Classification.1','Variant_Classification.2'),with=F]
dd1 <- dd[,c('Tumor_Sample_Barcode','tm.1','Variant_Classification.1'),with=F]
dd2 <- dd[,c('Tumor_Sample_Barcode','tm.2','Variant_Classification.2'),with=F]
names(dd1) <- c('Tumor_Sample_Barcode','tm','Variant_Classification')
dd1$allele <- 1
names(dd2) <- c('Tumor_Sample_Barcode','tm','Variant_Classification')
dd2$allele <- 2
dd <- rbind(dd1, dd2)
dd <- dd[tm %in% tmp$query_mutation,]
get_prop_missense <- function(dd) {
    n_mutants <- length(unique(strtrim(dd$Tumor_Sample_Barcode,9)))
    n_missense_mutants <- length(unique(strtrim(dd$Tumor_Sample_Barcode[dd$Variant_Classification=='Missense_Mutation'],9)))
    percentage_missense <- n_missense_mutants / n_mutants
    list(percentage_missense=percentage_missense)
}
res <- dd[,get_prop_missense(.SD),by=tm]
setnames(res,'percentage_missense','Prop. Missense')
out <- merge(out, res, by.x='query_mutation', by.y='tm', all.x=T)
setnames(out,'query_mutation','Residue')
setnames(out,'q.value.enriched','Q-value')
out <- out[order(`Q-value`,decreasing=F),]
out
```


## Paragraph 10

### Percentage of mutations including common TERT promoter variants that are composite

```{r, echo=T}
d <- d_mutations
d <- d[Hugo_Symbol=='TERT' & Variant_Classification=='TERT promoter' & exclude==F,]
get_tert_mutation <- function(Start_Position,Reference_Allele,Tumor_Seq_Allele2) {
    Amino_Acid_Position <- as.integer(substr(Start_Position,5,7))
    Reference_Allele <- toupper(Reference_Allele)
    Tumor_Seq_Allele2 <- toupper(Tumor_Seq_Allele2)
    mutation <- paste0(Amino_Acid_Position,'[',Reference_Allele,'>',Tumor_Seq_Allele2,']')
    mutation
}
d$mutation <- get_tert_mutation(d$Start_Position, d$Reference_Allele, d$Tumor_Seq_Allele2)
mutations <- unique(d$mutation)

getN <- function(query.mutation, d) {
    mutant_samples <- d$Tumor_Sample_Barcode[d$mutation==query.mutation]
    samples_with_mutation <- d$Tumor_Sample_Barcode[d$Tumor_Sample_Barcode %in% mutant_samples]
    tbl <- table.freq(samples_with_mutation)
    composite_samples <- tbl$value[tbl$N > 1]
    singleton_samples <- tbl$value[tbl$N == 1]
    n_composite <- length(unique(strtrim(composite_samples, 9)))
    n_singleton <- length(unique(strtrim(singleton_samples, 9)))
    n_overall <- length(unique(strtrim(tbl$value, 9)))
    list(mutation=query.mutation,n_composite=n_composite, n_singleton=n_singleton, n_overall=n_overall)
}
l <- lapply(mutations, getN, d)
l <- rbindlist(l)
info <- l[order(n_overall,decreasing=T),]
info <- info[n_overall >= 5,]
info$`Percentage as composites` <- round(100*info$n_composite / info$n_overall,3)
info[,n_singleton:=NULL]
info$mutation <- paste('TERT',info$mutation)
info
```






# Methods


### Number of tumor specimens in study
```{r, echo=T}
n_samples
```

### Number of distinct cancer subtypes
```{r, echo=T}
length(unique(d_clinical$CANCER_TYPE))
```

### Number of samples in TCGA data
```{r, echo=T}
length(unique(d_clinical_tcga$Tumor_Sample_Barcode))
```

### Number of cancer types in study
```{r, echo=T}
length(unique(d_clinical$metamaintype))
```



# relocated text:

### Number of samples in TCGA used in analysis
```{r, echo=T}
n_samples_tcga <- length(unique(d_clinical_tcga$Tumor_Sample_Barcode))
n_samples_tcga
```

### Percentage of TCGA samples with any composite mutations in an IMPACT gene
```{r, echo=T}
summarize_sample <- function(d) {
    tbl0 <- table.freq(d$Hugo_Symbol)
    any_composite <- any(tbl0$N > 1)
    list(any_composite=any_composite)
}

impact_genes <- gene_info$Hugo_Symbol
result <- d_mutations_tcga[Hugo_Symbol %in% impact_genes,summarize_sample(.SD),by=Tumor_Sample_Barcode]
n_samples_with_composites_tcga <- sum(result$any_composite)
round(100*(n_samples_with_composites_tcga / n_samples_tcga),3)
```




### Percentage of missense mutations in composites that are hotspots
```{r, echo=T}
#composites <- dd[composite==T & Variant_Classification=='Missense_Mutation',]
#composite_hotspots <- sum(composites$hotspot_class %in% c('24k','24k+3d'))
#n_composites <- nrow(composites)
#composite_hotspots/n_composites
```

### Percentage of singleton missense mutations that are hotspots
```{r, echo=T}
#singletons <- dd[composite==F & Variant_Classification=='Missense_Mutation',]
#singleton_hotspots <- sum(singletons$hotspot_class %in% c('24k','24k+3d'))
#n_singletons <- nrow(singletons)
#singleton_hotspots/n_singletons
```

### P-value for difference in proportion of hotspots in composites vs singletons
```{r, echo=T}
#tbl_composite <- table(composites$hotspotid!='')
#tbl_singleton <- table(singletons$hotspotid!='')
#x <- c(tbl_composite[['TRUE']], tbl_singleton[['TRUE']])
#n <- c(nrow(composites),nrow(singletons))
#pval <- prop.test(x,n,alternative='two.sided')$p.value
#pval
```

### Preference for cis vs trans in oncogenes vs TSGs does not exist for composites with silent mutations
```{r, echo=T}
#test_phase_role_preference <- function(d) {
#    d <- d[exclude==F & putative_resistance_mutation.1==F & putative_resistance_mutation.2==F,]
#    ## specifically require knowledge of clonalities so that we minimize cis bias
#    d <- d[same_cell_known==T & !is.na(ccf_Mcopies.1) & !is.na(ccf_Mcopies.2),]
#    d <- d[phase %in% c('cis','trans') & Role %in% c('Oncogene','TSG')]
#    x <- as.data.frame.matrix(xtabs(~phase + Role, data=d))
#    trans <- as.integer(x[1,])
#    phaseable <- colSums(x)
#    prop.test(trans, phaseable)$p.value
#}
#
## test for cis/trans preference between Onc/TSGs among composite mutations with non-synonymous variants
#test_phase_role_preference(d_phased)
#
## repeat for composites with 1+ silent mutations
#d_phased_silent <- fread(here('data/data_mutations_phased_withsilent.txt.gz'))
#test_phase_role_preference(d_phased_silent)

```

