---
title: "Compound-mutations: Values from data referenced in text"
output: html_document
chunk_output_type: console
---

```{r setup, include=F, echo=F}
knitr::opts_chunk$set(echo = F, message = F, warning = FALSE, autodep = knitr::dep_prev())
source(here::here('r/prerequisites.R'))
d_clinical <- fread(here('data/data_clinical.txt'))
d_mutations <- fread(here('data/data_mutations.txt'),select=c('Tumor_Sample_Barcode','exclude','Hugo_Symbol','Role','putative_resistance_mutation',
                                                              'Variant_Classification','Variant_Type','tcn', 'mutationid', 'high_tmb','hotspot_class',
                                                              'tm','hotspotid','HGVSp_Short'))
d_clinical_tcga <- fread(here('data/data_clinical_tcga.txt'))
d_mutations_tcga <- fread(here('data/data_mutations_tcga.txt'),select=c('Tumor_Sample_Barcode','Hugo_Symbol'))
gene_info <- fread(here('data/gene_info.txt'))
d_phased <- fread(here('data/data_mutations_phased.txt')) 
```

### Abstract

#### Number of patients in study:

```{r, echo=T}
length(unique(strtrim(d_clinical$Tumor_Sample_Barcode,9)))
```

#### Percentage of tumors harboring a compound mutation in an IMPACT gene:

```{r, echo=T}
n_samples <- length(unique(d_clinical$Tumor_Sample_Barcode))
check_compounds_per_sample <- function(d) {
    any_compound <- any(duplicated(d$Hugo_Symbol))
    list(any_compound=any_compound)
}
result <- d_mutations[,check_compounds_per_sample(.SD),by=Tumor_Sample_Barcode]
samples_with_compounds <- result$Tumor_Sample_Barcode[result$any_compound==T]
n_samples_with_compounds <- length(unique(samples_with_compounds))
fraction_samples_with_compounds <- n_samples_with_compounds / n_samples
round(100*fraction_samples_with_compounds)
```

### Paragraph 2

#### Number of cancer types in study
```{r, echo=T}
length(unique(d_clinical$metamaintype))
```

#### Number of tumor specimens in study
```{r, echo=T}
n_samples
```

#### Number of tumor specimens harboring any compound mutations
```{r, echo=T}
n_samples_with_compounds
```

#### Expected percentage of tumor specimens harboring any compound mutations by chance
```{r, echo=T}
## load the percent of compounds across all samples after each of 100,000 random permutations
## the average is the expected percentage of compound mutations by chance
l <- readRDS(here('data/observed_vs_expected_compounds_impact.rds'))
average_expected <- mean(l$dat_all$prop)
average_expected
```

#### Percent increase of observed to expected rate of compound-mutated samples in data
```{r, echo=T}
round(100*(fraction_samples_with_compounds / average_expected - 1),3)
```

#### Mutation burdens with enriched compound mutations
```{r,echo=T}
l <- readRDS(here('data/observed_vs_expected_compounds_per_tmb_impact.rds'))
get_p <- function(x) {
    pvalue <- x$p
    tmb <- x$burden
    list(tmb=tmb,pvalue=pvalue)
}
result <- rbindlist(lapply(l, get_p))
result[pvalue < 1e-5,]
```

#### Percentage of all samples with TMB: 4-12
```{r, echo=T}
x <- sum(round(d_clinical$tmb) %in% 4:12)
round(100*(x/n_samples),3)
```

#### Percetange of low TMB tumors with any compounds
```{r, echo=T}
low_tmb_samples <- d_clinical$Tumor_Sample_Barcode[round(d_clinical$tmb) == 4]
low_tmb_samples_with_compound <- samples_with_compounds[samples_with_compounds %in% low_tmb_samples]
round(100*length(unique(low_tmb_samples_with_compound)) / length(unique(low_tmb_samples)))
```

### Paragraph 3

#### Percentage of compound-mutant cases that could be ascribed to known phenomena

```{r, echo=T}
## categorize all samples in clinical data (including those with 0 mutations in MAF) according to cause of hypermutation or non-hypermutated
pol_samples <- sortunique(d_clinical$Tumor_Sample_Barcode[d_clinical$pol_signature==T & !is.na(d_clinical$pol_signature)])
msi_samples <- sortunique(d_clinical$Tumor_Sample_Barcode[d_clinical$msi_high==T & !is.na(d_clinical$msi_high)])
tmz_samples <- sortunique(d_clinical$Tumor_Sample_Barcode[d_clinical$tmz_signature==T & !is.na(d_clinical$tmz_signature)])
mmr_samples <- sortunique(d_clinical$Tumor_Sample_Barcode[d_clinical$mmr_signature==T & !is.na(d_clinical$mmr_signature)])

total_exclude <- sortunique(c(pol_samples, msi_samples, tmz_samples, mmr_samples))
normal_tmb_samples <- d_clinical$Tumor_Sample_Barcode[d_clinical$high_tmb==F & !is.na(d_clinical$high_tmb)]
normal_tmb_samples <- normal_tmb_samples[normal_tmb_samples %nin% total_exclude]
high_tmb_samples <- d_clinical$Tumor_Sample_Barcode[d_clinical$high_tmb==T & !is.na(d_clinical$high_tmb)]
other_high_tmb_samples <- high_tmb_samples[high_tmb_samples %nin% total_exclude]
tmp1 <- data.table(sample=normal_tmb_samples,group='Normal-TMB')
tmp2 <- data.table(sample=pol_samples,group='POLE-signature')
tmp3 <- data.table(sample=msi_samples,group='MSI-High')
tmp4 <- data.table(sample=tmz_samples,group='TMZ-signature')
tmp5 <- data.table(sample=mmr_samples,group='MMR-signature')
tmp6 <- data.table(sample=other_high_tmb_samples,group='Non-hypermutant, High-TMB')
tmp <- rbind(tmp1,tmp2,tmp3,tmp4,tmp5,tmp6)
tmp$group <- factor(tmp$group,levels=unique(tmp$group))
tmp <- tmp[order(sample,group),]
tmp <- tmp[!duplicated(sample),]

## annotate each sample/gene combo as compound, resistance, or probable biallelic loss
summarize_sample_gene <- function(d) {
    compound <- nrow(d) > 1
    any_resistance <- any(d$putative_resistance_mutation,na.rm=T) & compound==T
    if(d$Role[1] %in% c('TSG','Oncogene/TSG') & sum(d$Variant_Classification %in% c('Nonsense_Mutation','Splice_Site','Frame_Shift_Del','Frame_Shift_Ins')) > 1) {
        prob_biallelic_loss_of_tsg <- T
    } else {
        prob_biallelic_loss_of_tsg <- F
    }

    list(compound=compound,any_resistance=any_resistance,prob_biallelic_loss_of_tsg=prob_biallelic_loss_of_tsg)
}
info <- d_mutations[,summarize_sample_gene(.SD),by=c('Tumor_Sample_Barcode','Hugo_Symbol')]
info2 <- merge(info, tmp, by.x='Tumor_Sample_Barcode', by.y='sample', all=T)
info2[is.na(compound),compound:=F]
info2[is.na(any_resistance),any_resistance:=F]
info2[is.na(prob_biallelic_loss_of_tsg),prob_biallelic_loss_of_tsg:=F]

## summarize each sample for any compound mutations and their possible mechanisms
summarize_sample <- function(d) {
    any.compound <- any(d$compound)
    any.resistance <- any(d$any_resistance)
    if(any.compound==F) {
        reason <- 'n/a'
    } else {
        reason <- 'Unexplained compound mutation'
        if(d$group[1]!='Normal-TMB') {
            reason <- as.character(d$group[1])
        } else if(any.resistance) {
            reason <- 'Known resistance mutation'
        } else if(any(d$prob_biallelic_loss_of_tsg==T)) {
            reason <- 'Probably biallelic-loss of TSG'
        }
    }
    list(any.compound=any.compound, reason=reason)
}
sample_result <- info2[,summarize_sample(.SD),by=Tumor_Sample_Barcode]
sample_result <- sample_result[reason!='n/a']

## % of compound-mutant cases that could be ascribed to known phenomena
round(100*sum(sample_result$reason!='Unexplained compound mutation') / nrow(sample_result),3)
```

#### Percentage of compound-mutant cases attributable to resistance mutations
```{r, echo=T}
round(100*sum(sample_result$reason=='Known resistance mutation') / nrow(sample_result),3)
```  

#### Number of samples in TCGA used in analysis
```{r, echo=T}
n_samples_tcga <- length(unique(d_clinical_tcga$Tumor_Sample_Barcode))
n_samples_tcga
```

#### Percentage of TCGA samples with any compound mutations in an IMPACT gene
```{r, echo=T}
summarize_sample <- function(d) {
    tbl0 <- table.freq(d$Hugo_Symbol)
    any_compound <- any(tbl0$N > 1)
    list(any_compound=any_compound)
}

impact_genes <- gene_info$Hugo_Symbol
result <- d_mutations_tcga[Hugo_Symbol %in% impact_genes,summarize_sample(.SD),by=Tumor_Sample_Barcode]
n_samples_with_compounds_tcga <- sum(result$any_compound)
round(100*(n_samples_with_compounds_tcga / n_samples_tcga),3)
```

#### Percentage of compound-mutant cases attributable to putative biallelic-loss of TSGs
```{r, echo=T}
round(100*sum(sample_result$reason=='Probably biallelic-loss of TSG') / nrow(sample_result),3)
```  


#### Percentage of mutations in oncogenes and TSGs that are compound-mutants
```{r, echo=T}
d <- d_mutations[exclude==F & putative_resistance_mutation==F,]
d$id <- paste(d$Tumor_Sample_Barcode,d$Hugo_Symbol)
tbl <- table.freq(d$id)
tbl$compound <- tbl$N > 1
tbl[,N:=NULL]
d <- merge(d, tbl, by.x='id', by.y='value', all.x=T)

## Percentage of oncogene-mutant samples with a compound mutation
d_onc <- d[Role=='Oncogene',]
n_samples_onc <- length(unique(strtrim(d_onc$Tumor_Sample_Barcode,9)))
n_compound_onc <- length(unique(strtrim(d_onc$Tumor_Sample_Barcode[d_onc$compound==T],9)))
round(100*n_compound_onc/n_samples_onc,3)

## tabulate fraction of tsg-mutant samples with a compound mutation
d_tsg <- d[Role=='TSG',]
n_samples_tsg <- length(unique(strtrim(d_tsg$Tumor_Sample_Barcode,9)))
n_compound_tsg <- length(unique(strtrim(d_tsg$Tumor_Sample_Barcode[d_tsg$compound==T],9)))
round(100*n_compound_tsg/n_samples_tsg,3)

## p-value based on two-sample Z test 
tst <- prop.test(c(n_compound_onc,n_compound_tsg),c(n_samples_onc,n_samples_tsg))
pval <- tst$p.value
pval
```


#### Percentage of compound mutations in TSGs or Oncogenes with 1+ truncating variants vs only missense
```{r, echo=T}

## load phased data with pairwise-mutation comparison
d <- d_phased
d <- d[exclude==F & putative_resistance_mutation.1==F & putative_resistance_mutation.2==F,]
d$Class <- ''
d[truncating.1==F & truncating.2==F,Class:='Non-truncating/Non-truncating']
d[(truncating.1==T & truncating.2==F | truncating.1==F & truncating.2==T),Class:='Non-truncating/Truncating']
d[truncating.1==T & truncating.2==T,Class:='Truncating/Truncating']
d <- d[Role %in% c('Oncogene','TSG'),]

## don't overcount 3+ mutation compound-mutations by using unique sample/gene/Class counts
d$id <- paste(d$Tumor_Sample_Barcode, d$Hugo_Symbol, d$Class)
d <- d[!duplicated(id),c('Tumor_Sample_Barcode','Hugo_Symbol','Class','Role','id'),with=F]
dd <- d
dd[Class!='Non-truncating/Non-truncating',Class:='1 or both are truncating']
dd$Class <- factor(dd$Class,levels=(c('Non-truncating/Non-truncating','1 or both are truncating')))

## for each gene, get its singleton/compound counts
summarize_gene <- function(d) {
    n_missense_missense <- length(unique(strtrim(d$Tumor_Sample_Barcode[d$Class=='Non-truncating/Non-truncating'],9)))
    n_any_truncating <- length(unique(strtrim(d$Tumor_Sample_Barcode[d$Class=='1 or both are truncating'],9)))
    tbl <- list(`Non-truncating/Non-truncating`=n_missense_missense, `1 or both are truncating`=n_any_truncating)
    tbl
}
res <- dd[,summarize_gene(.SD),by=c('Role')]
m <- as.matrix(res[,2:3,with=F])
rownames(m) <- res$Role
round(100*(m / rowSums(m)),3)
```

#### Percentage of all affected cases that could not be ascribed to a known mutational mechanism or origin
```{r, echo=T}
round(100*sum(sample_result$reason=='Unexplained compound mutation') / nrow(sample_result),3)
```

### Paragraph 4

#### Genes significantly enriched for compound mutations

```{r, echo=T}
d <- d_mutations
d <- d[exclude==F & putative_resistance_mutation==F,]
d <- merge(d, gene_info, by='Hugo_Symbol', all.x=T)

## consider the TERT promoter as a unique gene, where the cds_length is now the total length of the promoter region covered by IMPACT
tert_promoter <- which(d$Hugo_Symbol=='TERT' & d$Variant_Classification=='TERT promoter')
d[tert_promoter, Hugo_Symbol:='TERT promoter']
d[tert_promoter, cds_length:=499] ## this is the length of the promoter region of TERT covered by IMPACT468
d[tert_promoter, percentage_gene_gc_content:=0.7896] ## this is the %GC content of TERT promoter region

## get observed number of compound mutated samples and mutated samples per gene
get_compound_rate_per_gene <- function(d) {
    tbl <- table.freq(d$Tumor_Sample_Barcode)
    samples <- tbl$value
    compound_samples <- tbl$value[tbl$N > 1]
    singleton_samples <- tbl$value[tbl$N == 1]

    ## don't overcount multiple samples per patient
    samples <- length(unique(strtrim(samples, 9)))
    compounds <- length(unique(strtrim(compound_samples, 9)))
    singletons <- length(unique(strtrim(singleton_samples, 9)))
    tcn <- mean(d$tcn,na.rm=T) 

    list(singletons=singletons,compounds=compounds,samples=samples,tcn=tcn,
         reptime=unique(d$reptime),percentage_gene_gc_content=unique(d$percentage_gene_gc_content),cds_length=unique(d$cds_length),
         panel_introduced=unique(d$panel_introduced))
}
res <- d[,get_compound_rate_per_gene(.SD),by=Hugo_Symbol]
res$panel_introduced <- factor(res$panel_introduced)
res$Hugo_Symbol <- gsub('-','_',res$Hugo_Symbol)
res$Hugo_Symbol <- gsub(' ','_',res$Hugo_Symbol)

## use negative-binomial regression to model the predicted number of compound-mutant samples per gene by chance
m <- glm.nb(compounds ~ offset(log(samples)) + reptime + cds_length + percentage_gene_gc_content + panel_introduced + tcn, data=res)
predictions <- predict(m,newdata=res,type='response')
res$predicted_proportion <- predictions / res$samples
res$observed_proportion <- res$compounds / res$samples
dat <- res[,c('Hugo_Symbol','compounds','samples','predicted_proportion','observed_proportion'),with=F]

## test each gene for enriched or depleted compound mutant samples
test_gene <-function(dat) {
    x <- tryCatch({
        tst <- binom.test(dat$compounds,dat$samples,dat$predicted_proportion,alternative='greater')
        p_enriched <- tst$p.value
        tst2 <- binom.test(dat$compounds,dat$samples,dat$predicted_proportion,alternative='two.sided')
        p_twosided <- tst2$p.value
        list(p_enriched=p_enriched, p_twosided=p_twosided)
    },error=function(e) {
        list(p_enriched=as.numeric(NA), p_twosided=as.numeric(NA))
    })
    dat$p_enriched <- x$p_enriched
    dat$p_twosided <- x$p_twosided
    dat
}
dat <- dat[,test_gene(.SD),by=Hugo_Symbol]
dat$logOR <- log2(dat$observed_proportion/dat$predicted_proportion)
x <- gene_info[,c('Hugo_Symbol','Role'),with=F]
x$Hugo_Symbol <- gsub('-','_',x$Hugo_Symbol)
dat <- merge(dat, x,by='Hugo_Symbol', all.x=T)
dat <- dat[order(p_enriched,decreasing=F),]
dat$q_enriched <- p.adjust(dat$p_enriched,method='BH')
dat$q_twosided <- p.adjust(dat$p_twosided,method='BH')

## genes with significant enrichment for compound mutations 
dat[q_enriched < 0.01,c('Hugo_Symbol','Role','q_enriched'),with=F]
```

#### Percentage of PIK3CA mutations that are compound mutations
```{r, echo=T}
## % of all pik3ca mutations that are compounds
x <- d_mutations[exclude==F & Hugo_Symbol=='PIK3CA',c('Tumor_Sample_Barcode','Hugo_Symbol'),with=F]
tbl <- table.freq(x$Tumor_Sample_Barcode)
samples_overall <- length(unique(strtrim(tbl$value,9)))
compounds <- length(unique(strtrim(tbl$value[tbl$N > 1],9)))
ci <- binom.confint(compounds, samples_overall, method='exact')
out <- adt(ci[,c('mean','lower','upper')])
out <- 100*out
names(out) <- c('Percentage','95%CI (lower)','95%CI (upper)')
out
``` 

#### Genes depleted for compound mutations
```{r, echo=T}
dat[q_twosided < 0.01 & logOR < 0,c('Hugo_Symbol','Role','q_twosided','logOR'),with=F]
```

#### IDH1 is depleted for compound mutations
```{r, echo=T}
dat[Hugo_Symbol=='IDH1',c('Hugo_Symbol','Role','q_twosided','logOR'),with=F]
```


### Paragraph 5

#### How many times more likely are compound mutations to include hotspots than singleton mutations?
```{r, echo=T}

## load MAF, subset for hotspot SNPs among valid samples
d <- d_mutations
d <- d[exclude==F & putative_resistance_mutation==F,]
dd <- d[Variant_Classification %in% 'Missense_Mutation',]
dd[hotspot_class=='3d',hotspot:=F]
dd[hotspot_class=='3d',hotspot_class:='']

## annotate if the sample-gene is singleton or compound
dd$id <- paste(dd$Tumor_Sample_Barcode,dd$Hugo_Symbol)
tbl <- table.freq(dd$id)
singleton <- tbl$value[tbl$N==1]
compound <- tbl$value[tbl$N>1]
dd$compound <- dd$id %in% compound
dd$PATIENT_ID <- strtrim(dd$Tumor_Sample_Barcode,9)
dd$pid_gene <- paste(dd$PATIENT_ID,dd$Hugo_Symbol)
total_number_singleton_mutations <- length(unique(dd$pid_gene[dd$compound==F]))
total_number_compound_mutations <- length(unique(dd$pid_gene[dd$compound==T]))
total_number_mutations <- length(unique(dd$pid_gene))

## for each hotspot, count the numnber of mutant, singleton and compound samples
summarize_individual_hotspot <- function(d) {
    ## count affected singleton/compound samples
    mutant_samples <- sortunique(d$Tumor_Sample_Barcode)
    singleton_samples <- sortunique(d$Tumor_Sample_Barcode[d$compound==F])
    compound_samples <- sortunique(d$Tumor_Sample_Barcode[d$compound==T])

    ## don't overcount patients with 2+ samples
    mutant_samples <- unique(strtrim(mutant_samples,9))
    singleton_samples <- unique(strtrim(singleton_samples,9))
    compound_samples <- unique(strtrim(compound_samples,9))

    n.mutant <- length(mutant_samples)
    n.singletons <- length(singleton_samples)
    n.compounds <- length(compound_samples)
    list(n.mutant=n.mutant,n.singletons=n.singletons,n.compounds=n.compounds)
}

hotspots <- sortunique(dd$hotspotid)
hotspots <- hotspots[hotspots!='']
ll <- dd[hotspotid!='',summarize_individual_hotspot(.SD),by=hotspotid]
ll <- ll[order(n.mutant,decreasing=T),]
ll$x.order <- 1:nrow(ll)
ll$p.singletons <- cumsum(ll$n.singletons / total_number_singleton_mutations)
ll$p.compounds <- cumsum(ll$n.compounds / total_number_compound_mutations)
ll$p.mutant <- cumsum(ll$n.mutant / total_number_mutations)

overall_prob_compound_mutation_has_hotspot <- max(ll$p.compounds)
overall_prob_singleton_mutation_has_hotspot <- max(ll$p.singleton)
overall_prob_compound_mutation_has_hotspot / overall_prob_singleton_mutation_has_hotspot

## Calculate P-value:
singletons_summary <- data.table(x=sum(ll$n.singletons),n=total_number_singleton_mutations,class='Singletons')
compounds_summary <- data.table(x=sum(ll$n.compounds),n=total_number_compound_mutations,class='Compound mutations')
summary <- rbind(singletons_summary, compounds_summary)
prop.test(summary$x, summary$n)$p.value
```

#### Percentage of singleton missense mutations that are hotspots
```{r, echo=T}
singletons <- dd[compound==F & Variant_Classification=='Missense_Mutation',]
singleton_hotspots <- sum(singletons$hotspot_class %in% c('24k','24k+3d'))
n_singletons <- nrow(singletons)
singleton_hotspots/n_singletons
```

#### Percentage of missense mutations in compounds that are hotspots
```{r, echo=T}
compounds <- dd[compound==T & Variant_Classification=='Missense_Mutation',]
compound_hotspots <- sum(compounds$hotspot_class %in% c('24k','24k+3d'))
n_compounds <- nrow(compounds)
compound_hotspots/n_compounds
```

#### P-value for difference in proportion of hotspots in compounds vs singletons
```{r, echo=T}
tbl_compound <- table(compounds$hotspotid!='')
tbl_singleton <- table(singletons$hotspotid!='')
x <- c(tbl_compound[['TRUE']], tbl_singleton[['TRUE']])
n <- c(nrow(compounds),nrow(singletons))
pval <- prop.test(x,n,alternative='two.sided')$p.value
pval
```


#### Percentage of compound mutations per gene role where more-prevalent allele is mutated first
```{r, echo=T}

## load phasing data; annotate functional mutations
d <- d_phased
d <- d[putative_resistance_mutation.1==F & putative_resistance_mutation.2==F & exclude==F,]
d[hotspotid.1!='',tm.1:=hotspotid.1] ## this will group IF-INDEL hotspots
d[hotspotid.2!='',tm.2:=hotspotid.2]
d$functional.1 <- d$hotspotid.1!='' | d$oncogenic.1 %in% c('Oncogenic','Likely Oncogenic','Predicted Oncogenic') |
(d$Role %in% c('TSG','Oncogene/TSG') & d$truncating.1==T) | d$Variant_Classification.1=='TERT promoter'
d$functional.2 <- d$hotspotid.2!='' | d$oncogenic.2 %in% c('Oncogenic','Likely Oncogenic','Predicted Oncogenic') |
(d$Role %in% c('TSG','Oncogene/TSG') & d$truncating.2==T) | d$Variant_Classification.2=='TERT promoter'
d[Variant_Classification.1 %in% c('TERT promoter'),tm.1:=paste0('TERT ',gsub('p[.]','',HGVSp_Short.1))]
d[Variant_Classification.1 %in% c('Splice_Site','Nonsense_Mutation','Frame_Shift_Del','Frame_Shift_Ins'), tm.1:=paste(Hugo_Symbol,'Truncating')]
d[Variant_Classification.2 %in% c('TERT promoter'),tm.2:=paste0('TERT ',gsub('p[.]','',HGVSp_Short.2))]
d[Variant_Classification.2 %in% c('Splice_Site','Nonsense_Mutation','Frame_Shift_Del','Frame_Shift_Ins'), tm.2:=paste(Hugo_Symbol,'Truncating')]
d <- d[,c('Tumor_Sample_Barcode','Role','Hugo_Symbol','hotspot.1','hotspot.2','functional.1','functional.2',
        'tm.1','tm.2','phase','common_reads_alt1_alt2',
        'common_reads_alt1_ref2','common_reads_ref1_alt2','ccf_Mcopies_lower.1','ccf_Mcopies_upper.1',
        'ccf_Mcopies_lower.2','ccf_Mcopies_upper.2','Variant_Classification.1','Variant_Classification.2','phase','same_cell_known'),with=F]

## annotate which allele was mutated first where this can be determined
d$firstallele <- ''
d$secondallele <- ''
d[ccf_Mcopies_lower.1 > ccf_Mcopies_upper.2,firstallele:=tm.1]
d[ccf_Mcopies_lower.1 > ccf_Mcopies_upper.2,secondallele:=tm.2]
d[ccf_Mcopies_upper.1 < ccf_Mcopies_lower.2,firstallele:=tm.2]
d[ccf_Mcopies_upper.1 < ccf_Mcopies_lower.2,secondallele:=tm.1]
dd <- d[,c('Tumor_Sample_Barcode','Role','Hugo_Symbol','firstallele','secondallele','phase','same_cell_known','hotspot.1','hotspot.2',
        'functional.1','functional.2','Variant_Classification.1','Variant_Classification.2'),with=F]
dd <- dd[firstallele!='' & secondallele!='',]

## get number of unique patients with each mutation
d0 <- d_mutations
d0 <- d0[exclude==F & putative_resistance_mutation==F,]
d0$functional <- d0$hotspot==T | d0$oncogenic %in% c('Oncogenic','Likely Oncogenic','Predicted Oncogenic') |
(d0$Role %in% c('TSG','Oncogene/TSG') & d0$truncating==T) | d0$Variant_Classification=='TERT promoter'
d0[Variant_Classification %in% c('TERT promoter'),tm:=paste0('TERT ',gsub('p[.]','',HGVSp_Short))]
d0[Variant_Classification %in% c('Splice_Site','Nonsense_Mutation','Frame_Shift_Del','Frame_Shift_Ins'), tm:=paste(Hugo_Symbol,'Truncating')]
d0[hotspotid!='',tm:=hotspotid] ## this will group IF-INDEL hotspots
getN <- function(d0) list(N=length(unique(strtrim(d0$Tumor_Sample_Barcode,9))))
tbl <- d0[,getN(.SD),by=tm]
setnames(tbl,'tm','value')

## annotate first and second hits with their prevalence across full dataset
dd2 <- merge(dd, tbl, by.x='firstallele', by.y='value', all.x=T)
setnames(dd2,'N','firstallele_N')
dd2 <- merge(dd2, tbl, by.x='secondallele', by.y='value', all.x=T)
setnames(dd2,'N','secondallele_N')
tbl_gene <- table.freq(d0$Hugo_Symbol)
dd2 <- merge(dd2, tbl_gene, by.x='Hugo_Symbol', by.y='value', all.x=T)
setnames(dd2,'N','totalN')

## compare prevalence of each hit for signficantly more prevalent:
first_ci <- binom::binom.confint(dd2$firstallele_N, dd2$totalN, method='exact')
second_ci <- binom::binom.confint(dd2$secondallele_N, dd2$totalN, method='exact')
dd2$prop_samples_with_first_allele <- first_ci$mean
dd2$prop_samples_with_first_allele_lwr <- first_ci$lower
dd2$prop_samples_with_first_allele_upr <- first_ci$upper
dd2$prop_samples_with_second_allele <- second_ci$mean
dd2$prop_samples_with_second_allele_lwr <- second_ci$lower
dd2$prop_samples_with_second_allele_upr <- second_ci$upper
dd2$more_common_allele <- 'no difference'
dd2[prop_samples_with_first_allele_lwr > prop_samples_with_second_allele_upr, more_common_allele:='first']
dd2[prop_samples_with_first_allele_upr < prop_samples_with_second_allele_lwr, more_common_allele:='second']
dd2$Hugo_Symbol <- as.character(dd2$Hugo_Symbol)
dd2$firstallele <- as.character(dd2$firstallele)
dd2$secondallele <- as.character(dd2$secondallele)
dat <- dd2

## get the prop of 1st more prevalent vs 2nd more prevalent vs gene role
dat <- dat[same_cell_known==T | phase=='cis',] ## only consider compounds definitely in same cell
dat$n_functional <- dat$functional.1 + dat$functional.2
dat <- dat[n_functional > 0,]
dat$n_functional <- paste0('functional_',dat$n_functional)
dat$n_functional <- factor(dat$n_functional)
dat <- dat[more_common_allele!='no difference',]
m <- as.data.frame(t(as.data.frame.matrix(xtabs(~more_common_allele + n_functional, data=dat[Role=='Oncogene']))))
onc_propdata <- adt(binom::binom.confint(m$first,(m$first+m$second),method='exact'))
onc_propdata$n_functional <- c(1,2)
onc_propdata$role <- 'Oncogenes'
m <- as.data.frame(t(as.data.frame.matrix(xtabs(~more_common_allele + n_functional, data=dat[Role=='TSG']))))
tsg_propdata <- adt(binom::binom.confint(m$first,(m$first+m$second),method='exact'))
tsg_propdata$n_functional <- c(1,2)
tsg_propdata$role <- 'TSGs'
propdata <- rbind(onc_propdata, tsg_propdata)
propdata <- propdata[n_functional==2,]
setnames(propdata,'mean','prop_first')
out <- propdata[,c('role','prop_first','lower','upper'),with=F]
out[,2:4] <- 100*out[,(2:4),with=F]
names(out) <- c('Role','% more prevalent mutated first','95%CI (lower','95%CI (upper)')

## Percentage of compound mutations per gene role where more-prevalent allele is mutated first
out
```


### Paragraph 6



